---
title: "limmde: Differential expression testing with linear mixed models for repeated measures"
author: "Gabriel E Hoffman"
date: "`r format(Sys.Date(), '%m/%d/%Y')`"
abstract: > 
  Perform powerfull test of differential expression while 
  cache ontroling the false positive rate  
  variancePartition package version: `r packageVersion("variancePartition")`
output:
  rmarkdown::html_document:
    highlight: pygments
    toc: true
    fig_width: 5
bibliography: library.bib
vignette: >
  %\VignetteIndexEntry{Differential expression testing with repeated measures}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\usepackage[utf8]{inputenc}
---

<!---
library(knitr)
rmarkdown::render('DESeq2.Rmd')
--->

```{r setup, echo=FALSE, results="hide"}
knitr::opts_chunk$set(tidy=FALSE, cache=TRUE,
                      dev="png",
                      message=FALSE, error=FALSE, warning=TRUE)
```	

# Standard workflow
**Note:** if you use DESeq2 in published research, please cite:

> Love, M.I., Huber, W., Anders, S. (2014)
> Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2.
> *Genome Biology*, **15**:550.
> [10.1186/s13059-014-0550-8](http://dx.doi.org/10.1186/s13059-014-0550-8)


```{r preprocess, eval=TRUE}
library(variancePartition)
library(limma)
library(edgeR)
library(doParallel)
data(varPartDEdata)

isexpr = rowSums(cpm(countMatrix)>0.1) >= 3

# Create VOBJ
genes = DGEList( countMatrix )
genes = calcNormFactors( genes )
design = model.matrix( ~ Disease, metadata)
vobj_tmp = voom( genes, design, plot=FALSE)
dupcor <- duplicateCorrelation(vobj_tmp,design,block=metadata$Individual)
vobj = voom( genes, design, plot=FALSE,block=metadata$Individual,correlation=dupcor$consensus)
```

Duplicate correlation
```{r dupCor, eval=TRUE}
design = model.matrix( ~ Disease, metadata)
dupcor <- duplicateCorrelation(vobj,design,block=metadata$Individual)
fitDupCor <- lmFit(vobj,design,block=metadata$Individual,correlation=dupcor$consensus)
fitDupCor <- eBayes(fitDupCor)

```

Lindear mixed model
```{r lmm, eval=TRUE}
cl <- makeCluster(4)
registerDoParallel(cl)

# fitMixedModelDE
form <- ~ Disease + (1|Individual) 
L = getContrast( vobj, form, metadata, "Disease1")
fitmm = fitMixedModelDE( vobj, form, metadata, L, showWarnings=FALSE)
fitmm_eb = eBayes( fitmm )
```




## Input data

### Why un-normalized counts?


# Session info
```{r sessionInfo}
sessionInfo()
```

# References

