<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>dream: Differential expression testing with linear mixed models for repeated measures • variancePartition</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="dream: Differential expression testing with linear mixed models for repeated measures">
<meta property="og:description" content="variancePartition">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">variancePartition</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.21.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/FAQ.html">Frequently asked questions</a>
    </li>
    <li>
      <a href="../articles/additional_visualization.html">Additional visualizations of variance structure</a>
    </li>
    <li>
      <a href="../articles/dream.html">dream: Differential expression testing with linear mixed models for repeated measures</a>
    </li>
    <li>
      <a href="../articles/theory_practice_random_effects.html">3) Theory and practice of random effects and REML</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/GabrielHoffman/variancePartition/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="dream_files/header-attrs-2.7/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>dream: Differential expression testing with linear mixed models for repeated measures</h1>
                        <h4 class="author"><a href="http://gabrielhoffman.github.io">Gabriel Hoffman</a></h4>
            <address class="author_afil">
      <p>Icahn School of Medicine at Mount Sinai, New York</p>
<br><small class="dont-index">Source: <a href="https://github.com/GabrielHoffman/variancePartition/blob/master/vignettes/dream.Rmd"><code>vignettes/dream.Rmd</code></a></small>
      <div class="hidden name"><code>dream.Rmd</code></div>

    </address>
</div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p><strong>D</strong>ifferential expression for <strong>re</strong>pe<strong>a</strong>ted <strong>m</strong>easures (dream) uses a linear model model to increase power and decrease false positives for RNA-seq datasets with multiple measurements per individual. The analysis fits seamlessly into the widely used workflow of limma/voom <span class="citation">(Law et al. 2014)</span>. <br><br> Dream uses a linear model model to increase power and decrease false positives for RNA-seq datasets with repeated measurements. Dream achieves this by combining multiple statistical concepts into a single statistical model. The model includes: <br><br>
      - flexible modeling of repeated measures gene expression data<br> - precision weights to model measurement error in RNA-seq counts<br> - ability to model multiple random effects<br> - random effects estimated separately for each gene<br> - hypothesis testing for fixed effects in linear mixed models<br> - small sample size hypothesis test<br><br> Dream also includes multi-threaded analysis across thousands of genes on a multi-core machine.<br><br> variancePartition v1.21.3<br> April 02, 2021 11:32:50<br></p>
    </div>
    
<!---
cd /Users/gabrielhoffman/workspace/repos
R
library('knitr')
rmarkdown::render('variancePartition/vignettes/dream.Rmd')

# create public version
grep -v "## Warning" /Users/gabrielhoffman/workspace/repos/variancePartition/vignettes/dream.html > /Users/gabrielhoffman/workspace/repos/variancePartition/vignettes/dream2.html
 scp /Users/gabrielhoffman/workspace/repos/variancePartition/vignettes/dream2.html minerva:/hpc/users/hoffmg01/www/software/dream.html
--->
<!---
# Description
Dream uses a linear model model to increase power and decrease false positives for RNA-seq datasets with repeated measurements.  Dream achieves this by combining multiple statistical concepts into a single statistical model.  The dream model includes:     


| Model property                                                 | Package           | Reference                           |
|:---------------------------------------------------------------|:------------------|:------------------------------------|
| precision weights to model measurement error in RNA-seq counts | limma/voom        | @Law2014                            |
| ability to model multiple random effects                       | lme4              | @Bates2015                          |
| random effects estimated separately for each gene              | variancePartition | @hoffman2016                        |
| hypothesis testing for fixed effects in linear mixed models    | lmerTest          | Kuznetsova, et al. -@kuznetsova2017 |
| small sample size hypothesis test                              | pbkrtest          | Halekoh, et al. -@halekoh2014       |
|                                                                |                   |                                     |
* flexible modeling of repeated measures gene expression data
* precision weights to model measurement error in RNA-seq counts
* ability to model multiple random effects
* random effects estimated separately for each gene
* hypothesis testing for fixed effects in linear mixed models
* small sample size hypothesis test
* multi-threaded analysis across thousands of genes on a multi-core machine.
--->
<div id="standard-rna-seq-processing" class="section level1">
<h1 class="hasAnchor">
<a href="#standard-rna-seq-processing" class="anchor"></a>Standard RNA-seq processing</h1>
<p>This tutorial assumes that the reader is familiar with the limma/voom workflow for RNA-seq. Process raw count data using limma/voom.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">'variancePartition'</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="http://bioinf.wehi.edu.au/edgeR">'edgeR'</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/Bioconductor/BiocParallel">'BiocParallel'</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">varPartDEdata</span><span class="op">)</span>

<span class="co"># filter genes by number of counts</span>
<span class="va">isexpr</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/cpm.html">cpm</a></span><span class="op">(</span><span class="va">countMatrix</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">0.1</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">5</span>

<span class="co"># Standard usage of limma/voom</span>
<span class="va">geneExpr</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html">DGEList</a></span><span class="op">(</span> <span class="va">countMatrix</span><span class="op">[</span><span class="va">isexpr</span>,<span class="op">]</span> <span class="op">)</span>
<span class="va">geneExpr</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html">calcNormFactors</a></span><span class="op">(</span> <span class="va">geneExpr</span> <span class="op">)</span>

<span class="co"># make this vignette faster by analyzing a subset of genes</span>
<span class="va">geneExpr</span> <span class="op">=</span> <span class="va">geneExpr</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">1000</span>,<span class="op">]</span></code></pre></div>
</div>
<div id="limma-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#limma-analysis" class="anchor"></a>Limma Analysis</h1>
<p>Limma has a built-in approach for analyzing repeated measures data using <code>duplicateCorrelation()</code>. The model can handle a single random effect, and forces the magnitude of the random effect to be the same across all genes.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># apply duplicateCorrelation is two rounds</span>
<span class="va">design</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span> <span class="op">~</span> <span class="va">Disease</span>, <span class="va">metadata</span><span class="op">)</span>
<span class="va">vobj_tmp</span> <span class="op">=</span> <span class="fu">voom</span><span class="op">(</span> <span class="va">geneExpr</span>, <span class="va">design</span>, plot<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="va">dupcor</span> <span class="op">&lt;-</span> <span class="fu">duplicateCorrelation</span><span class="op">(</span><span class="va">vobj_tmp</span>,<span class="va">design</span>,block<span class="op">=</span><span class="va">metadata</span><span class="op">$</span><span class="va">Individual</span><span class="op">)</span>

<span class="co"># run voom considering the duplicateCorrelation results</span>
<span class="co"># in order to compute more accurate precision weights</span>
<span class="co"># Otherwise, use the results from the first voom run</span>
<span class="va">vobj</span> <span class="op">=</span> <span class="fu">voom</span><span class="op">(</span> <span class="va">geneExpr</span>, <span class="va">design</span>, plot<span class="op">=</span><span class="cn">FALSE</span>, block<span class="op">=</span><span class="va">metadata</span><span class="op">$</span><span class="va">Individual</span>, correlation<span class="op">=</span><span class="va">dupcor</span><span class="op">$</span><span class="va">consensus</span><span class="op">)</span>

<span class="co"># Estimate linear mixed model with a single variance component</span>
<span class="co"># Fit the model for each gene, </span>
<span class="va">dupcor</span> <span class="op">&lt;-</span> <span class="fu">duplicateCorrelation</span><span class="op">(</span><span class="va">vobj</span>, <span class="va">design</span>, block<span class="op">=</span><span class="va">metadata</span><span class="op">$</span><span class="va">Individual</span><span class="op">)</span>

<span class="co"># But this step uses only the genome-wide average for the random effect</span>
<span class="va">fitDupCor</span> <span class="op">&lt;-</span> <span class="fu">lmFit</span><span class="op">(</span><span class="va">vobj</span>, <span class="va">design</span>, block<span class="op">=</span><span class="va">metadata</span><span class="op">$</span><span class="va">Individual</span>, correlation<span class="op">=</span><span class="va">dupcor</span><span class="op">$</span><span class="va">consensus</span><span class="op">)</span>

<span class="co"># Fit Empirical Bayes for moderated t-statistics</span>
<span class="va">fitDupCor</span> <span class="op">&lt;-</span> <span class="fu">eBayes</span><span class="op">(</span> <span class="va">fitDupCor</span> <span class="op">)</span></code></pre></div>
</div>
<div id="dream-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#dream-analysis" class="anchor"></a>Dream Analysis</h1>
<p>The dream method replaces two core functions of limma with a linear mixed model.</p>
<ol style="list-style-type: decimal">
<li>
<code><a href="../reference/voomWithDreamWeights.html">voomWithDreamWeights()</a></code> replaces <code>voom()</code> to estimate precision weights</li>
<li>
<code><a href="../reference/dream.html">dream()</a></code> replaces <code>lmFit()</code> to estimate regression coefficients.</li>
</ol>
<p>Otherwise dream uses the same workflow as limma with <code><a href="../reference/topTable.html">topTable()</a></code>, since any statistical differences are handled behind the scenes.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Specify parallel processing parameters</span>
<span class="co"># this is used implicitly by dream() to run in parallel</span>
<span class="va">param</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html">SnowParam</a></span><span class="op">(</span><span class="fl">4</span>, <span class="st">"SOCK"</span>, progressbar<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/register.html">register</a></span><span class="op">(</span><span class="va">param</span><span class="op">)</span>

<span class="co"># The variable to be tested must be a fixed effect</span>
<span class="va">form</span> <span class="op">&lt;-</span> <span class="op">~</span> <span class="va">Disease</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span><span class="op">|</span><span class="va">Individual</span><span class="op">)</span> 

<span class="co"># estimate weights using linear mixed model of dream</span>
<span class="va">vobjDream</span> <span class="op">=</span> <span class="fu"><a href="../reference/voomWithDreamWeights.html">voomWithDreamWeights</a></span><span class="op">(</span> <span class="va">geneExpr</span>, <span class="va">form</span>, <span class="va">metadata</span> <span class="op">)</span>

<span class="co"># Fit the dream model on each gene</span>
<span class="co"># By default, uses the Satterthwaite approximation for the hypothesis test</span>
<span class="va">fitmm</span> <span class="op">=</span> <span class="fu"><a href="../reference/dream.html">dream</a></span><span class="op">(</span> <span class="va">vobjDream</span>, <span class="va">form</span>, <span class="va">metadata</span> <span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Examine design matrix</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">fitmm</span><span class="op">$</span><span class="va">design</span>, <span class="fl">3</span><span class="op">)</span></code></pre></div>
<pre><code>##           (Intercept) Disease1
## sample_01           1        0
## sample_02           1        0
## sample_03           1        0</code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Get results of hypothesis test on coefficients of interest</span>
<span class="fu"><a href="../reference/topTable.html">topTable</a></span><span class="op">(</span> <span class="va">fitmm</span>, coef<span class="op">=</span><span class="st">'Disease1'</span>, number<span class="op">=</span><span class="fl">3</span> <span class="op">)</span></code></pre></div>
<pre><code>##                                    logFC  AveExpr        t      P.Value    adj.P.Val     z.std
## ENST00000283033.5 gene=TXNDC11  1.556233 3.567624 68.62592 3.701410e-27 3.701410e-24 10.793327
## ENST00000257181.9 gene=PRPF38A  1.380549 4.398270 35.53035 6.311837e-21 3.155919e-18  9.384662
## ENST00000421974.2 gene=ATP6V0E2 1.386821 3.478030 22.35157 1.289327e-16 3.584329e-14  8.274558</code></pre>
<p>Since dream uses an estimated degrees of freedom value for each hypothsis test, the degrees of freedom is different for each gene here. Therefore, the t-statistics are not directly comparable since they have different degrees of freedom. In order to be able to compare test statistics, we report <code>z.std</code> which is the p-value transformed into a signed z-score. This can be used for downstream analysis.</p>
<p>Note that if a random effect is not specified, <code><a href="../reference/dream.html">dream()</a></code> automatically uses <code>lmFit()</code>, but the user must run <code>eBayes()</code> afterward.</p>
<div id="advanced-hypothesis-testing" class="section level2">
<h2 class="hasAnchor">
<a href="#advanced-hypothesis-testing" class="anchor"></a>Advanced hypothesis testing</h2>
<div id="using-contrasts-to-compare-coefficients" class="section level3">
<h3 class="hasAnchor">
<a href="#using-contrasts-to-compare-coefficients" class="anchor"></a>Using contrasts to compare coefficients</h3>
<p>You can also perform a hypothesis test of the <em>difference</em> between two or more coefficients by using a contrast matrix. The contrasts are evaluated at the time of the model fit and the results can be extracted with <code><a href="../reference/topTable.html">topTable()</a></code>. This behaves like <code>makeContrasts()</code> and <code>contrasts.fit()</code> in limma.</p>
<p>Make sure to inspect your contrast matrix to confirm it is testing what you intend.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">form</span> <span class="op">&lt;-</span> <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">DiseaseSubtype</span> <span class="op">+</span> <span class="va">Sex</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span><span class="op">|</span><span class="va">Individual</span><span class="op">)</span>  
<span class="va">L</span> <span class="op">=</span> <span class="fu"><a href="../reference/getContrast.html">getContrast</a></span><span class="op">(</span> <span class="va">vobjDream</span>, <span class="va">form</span>, <span class="va">metadata</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"DiseaseSubtype2"</span>, <span class="st">"DiseaseSubtype1"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Visualize contrast matrix</span>
<span class="fu"><a href="../reference/plotContrasts.html">plotContrasts</a></span><span class="op">(</span><span class="va">L</span><span class="op">)</span> </code></pre></div>
<p><img src="dream_files/figure-html/contrast-1.png" width="672"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># fit dream model with contrasts</span>
<span class="va">fit</span> <span class="op">=</span> <span class="fu"><a href="../reference/dream.html">dream</a></span><span class="op">(</span> <span class="va">vobjDream</span>, <span class="va">form</span>, <span class="va">metadata</span>, <span class="va">L</span><span class="op">)</span>

<span class="co"># get names of available coefficients and contrasts for testing</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "L1"              "DiseaseSubtype0" "DiseaseSubtype1" "DiseaseSubtype2" "SexM"</code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># extract results from first contrast</span>
<span class="fu"><a href="../reference/topTable.html">topTable</a></span><span class="op">(</span> <span class="va">fit</span>, coef<span class="op">=</span><span class="st">"L1"</span>, number<span class="op">=</span><span class="fl">3</span> <span class="op">)</span></code></pre></div>
<pre><code>##                                       logFC  AveExpr         t      P.Value  adj.P.Val     z.std
## ENST00000355624.3 gene=RAB11FIP2 -0.9493146 5.260280 -5.384718 2.857043e-05 0.02857043 -4.184572
## ENST00000593466.1 gene=DDA1      -1.7265709 3.901579 -3.516848 2.168824e-03 0.99955525 -3.066084
## ENST00000200676.3 gene=CETP       1.4777422 3.723438  3.732332 8.085141e-03 0.99955525  2.648493</code></pre>
<p>Multiple contrasts can be evaluated at the same time, in order to save computation time:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">form</span> <span class="op">&lt;-</span> <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">DiseaseSubtype</span> <span class="op">+</span> <span class="va">Sex</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span><span class="op">|</span><span class="va">Individual</span><span class="op">)</span> 

<span class="co"># define and then cbind contrasts</span>
<span class="va">L1</span> <span class="op">=</span> <span class="fu"><a href="../reference/getContrast.html">getContrast</a></span><span class="op">(</span> <span class="va">vobjDream</span>, <span class="va">form</span>, <span class="va">metadata</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"DiseaseSubtype2"</span>, <span class="st">"DiseaseSubtype1"</span><span class="op">)</span><span class="op">)</span>
<span class="va">L2</span> <span class="op">=</span> <span class="fu"><a href="../reference/getContrast.html">getContrast</a></span><span class="op">(</span> <span class="va">vobjDream</span>, <span class="va">form</span>, <span class="va">metadata</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"DiseaseSubtype1"</span>, <span class="st">"DiseaseSubtype0"</span><span class="op">)</span><span class="op">)</span>
<span class="va">L</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/cbind.html">cbind</a></span><span class="op">(</span><span class="va">L1</span>, <span class="va">L2</span><span class="op">)</span>     

<span class="co"># Visualize contrasts</span>
<span class="fu"><a href="../reference/plotContrasts.html">plotContrasts</a></span><span class="op">(</span><span class="va">L</span><span class="op">)</span> </code></pre></div>
<p><img src="dream_files/figure-html/contrast.combine-1.png" width="672"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># fit both contrasts</span>
<span class="va">fit</span> <span class="op">=</span> <span class="fu"><a href="../reference/dream.html">dream</a></span><span class="op">(</span> <span class="va">vobjDream</span>, <span class="va">form</span>, <span class="va">metadata</span>, <span class="va">L</span><span class="op">)</span>

<span class="co"># extract results from first contrast</span>
<span class="fu"><a href="../reference/topTable.html">topTable</a></span><span class="op">(</span> <span class="va">fit</span>, coef<span class="op">=</span><span class="st">"L1"</span>, number<span class="op">=</span><span class="fl">3</span> <span class="op">)</span></code></pre></div>
<pre><code>##                                       logFC  AveExpr         t      P.Value  adj.P.Val     z.std
## ENST00000355624.3 gene=RAB11FIP2 -0.9493146 5.260280 -5.384718 2.857043e-05 0.02857043 -4.184572
## ENST00000593466.1 gene=DDA1      -1.7265709 3.901579 -3.516848 2.168824e-03 0.99955525 -3.066084
## ENST00000200676.3 gene=CETP       1.4777422 3.723438  3.732332 8.085141e-03 0.99955525  2.648493</code></pre>
</div>
<div id="comparing-multiple-coefficients" class="section level3">
<h3 class="hasAnchor">
<a href="#comparing-multiple-coefficients" class="anchor"></a>Comparing multiple coefficients</h3>
<p>So far contrasts have only involved the difference between two coefficients. But contrasts can also compare linear combinations of coefficients. A user can create contrasts ‘manually,’ as long as the elements sum to 0. Here, consider comparing <code>DiseaseSubtype0</code> to the mean of <code>DiseaseSubtype1</code> and <code>DiseaseSubtype2</code></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># the tests is DiseaseSubtype0 - (DiseaseSubtype1/2 + DiseaseSubtype2/2)</span>
<span class="co"># Note that the order of the coefficients must be the same as from getContrast()</span>
<span class="va">L3</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="op">-</span><span class="fl">1</span><span class="op">/</span><span class="fl">2</span>, <span class="op">-</span><span class="fl">1</span><span class="op">/</span><span class="fl">2</span>, <span class="fl">0</span><span class="op">)</span>

<span class="co"># combine L1 and L2 contrasts with manually defind L3 contrast.</span>
<span class="va">Lall</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/cbind.html">cbind</a></span><span class="op">(</span><span class="va">L</span>, <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>L3 <span class="op">=</span> <span class="va">L3</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Note that each contrast must sum to 0</span>
<span class="fu"><a href="../reference/plotContrasts.html">plotContrasts</a></span><span class="op">(</span><span class="va">Lall</span><span class="op">)</span></code></pre></div>
<p><img src="dream_files/figure-html/maual.contrasts-1.png" width="768"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># fit dream model to evaluate contrasts</span>
<span class="va">fit</span> <span class="op">=</span> <span class="fu"><a href="../reference/dream.html">dream</a></span><span class="op">(</span> <span class="va">vobjDream</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="op">]</span>, <span class="va">form</span>, <span class="va">metadata</span>, L<span class="op">=</span><span class="va">Lall</span><span class="op">)</span>

<span class="fu"><a href="../reference/topTable.html">topTable</a></span><span class="op">(</span><span class="va">fit</span>, coef<span class="op">=</span><span class="st">"L3"</span>, number<span class="op">=</span><span class="fl">3</span><span class="op">)</span></code></pre></div>
<pre><code>##                                     logFC  AveExpr         t      P.Value    adj.P.Val     z.std
## ENST00000456159.1 gene=MET     -0.9830788 2.458926 -6.349967 3.385843e-06 3.385843e-05 -4.645908
## ENST00000418210.2 gene=TMEM64  -1.0343236 4.715367 -8.899162 1.603627e-05 5.403575e-05 -4.313951
## ENST00000555834.1 gene=RPS6KL1 -0.8954794 5.272063 -5.636526 1.621072e-05 5.403575e-05 -4.311559</code></pre>
</div>
<div id="joint-hypothesis-test-of-multiple-coefficients" class="section level3">
<h3 class="hasAnchor">
<a href="#joint-hypothesis-test-of-multiple-coefficients" class="anchor"></a>Joint hypothesis test of multiple coefficients</h3>
<p>Joint hypothesis testing of multiple coefficients at the same time can be performed by using an F-test. Just like in limma, the results can be extracted using <code><a href="../reference/topTable.html">topTable()</a></code></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># extract results from first contrast</span>
<span class="fu"><a href="../reference/topTable.html">topTable</a></span><span class="op">(</span> <span class="va">fit</span>, coef<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"DiseaseSubtype2"</span>, <span class="st">"DiseaseSubtype1"</span><span class="op">)</span>, number<span class="op">=</span><span class="fl">3</span> <span class="op">)</span></code></pre></div>
<pre><code>##                                DiseaseSubtype2 DiseaseSubtype1  AveExpr         F      P.Value
## ENST00000418210.2 gene=TMEM64         5.301001        5.211674 4.715367 1330.8479 6.020760e-12
## ENST00000555834.1 gene=RPS6KL1        5.662699        5.719196 5.272063  846.4267 4.670490e-11
## ENST00000589123.1 gene=NFIC           6.545195        6.181023 5.855335  531.9997 3.802383e-10
##                                   adj.P.Val    F.std
## ENST00000418210.2 gene=TMEM64  6.020760e-11 25.83581
## ENST00000555834.1 gene=RPS6KL1 2.335245e-10 23.78717
## ENST00000589123.1 gene=NFIC    1.267461e-09 21.69022</code></pre>
<p>Since dream uses an estimated degrees of freedom value for each hypothsis test, the degrees of freedom is different for each gene here. Therefore, the F-statistics are not directly comparable since they have different degrees of freedom. In order to be able to compare test statistics, we report <code>F.std</code> which is the p-value transformed into an F-statistic with <span class="math inline">\(df_1={\text{number of coefficiets tested}}\)</span> and <span class="math inline">\(df_2=\infty\)</span>. This can be used for downstream analysis.</p>
</div>
</div>
<div id="small-sample-method" class="section level2">
<h2 class="hasAnchor">
<a href="#small-sample-method" class="anchor"></a>Small-sample method</h2>
<p>For small datasets, the Kenward-Roger method can be more powerful. But it is <strong>substantially</strong> more computationally intensive.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fitmmKR</span> <span class="op">=</span> <span class="fu"><a href="../reference/dream.html">dream</a></span><span class="op">(</span> <span class="va">vobjDream</span>, <span class="va">form</span>, <span class="va">metadata</span>, ddf<span class="op">=</span><span class="st">"Kenward-Roger"</span><span class="op">)</span></code></pre></div>
</div>
<div id="variancepartition-plot" class="section level2">
<h2 class="hasAnchor">
<a href="#variancepartition-plot" class="anchor"></a>variancePartition plot</h2>
<p>Dream and variancePartition share the same underlying linear mixed model framework. A variancePartition analysis can indicate important variables that should be included as fixed or random effects in the dream analysis.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Note: this could be run with either vobj from voom()</span>
<span class="co"># or vobjDream from voomWithDreamWeights()</span>
<span class="co"># The resuylts are similar</span>
<span class="va">form</span> <span class="op">=</span> <span class="op">~</span> <span class="op">(</span><span class="fl">1</span><span class="op">|</span><span class="va">Individual</span><span class="op">)</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span><span class="op">|</span><span class="va">Disease</span><span class="op">)</span>
<span class="va">vp</span> <span class="op">=</span> <span class="fu"><a href="../reference/fitExtractVarPartModel.html">fitExtractVarPartModel</a></span><span class="op">(</span> <span class="va">vobj</span>, <span class="va">form</span>, <span class="va">metadata</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotVarPart.html">plotVarPart</a></span><span class="op">(</span> <span class="fu"><a href="../reference/sortCols.html">sortCols</a></span><span class="op">(</span><span class="va">vp</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="dream_files/figure-html/vp-1.png" width="700"></p>
</div>
<div id="compare-p-values-from-dream-and-duplicatecorrelation" class="section level2">
<h2 class="hasAnchor">
<a href="#compare-p-values-from-dream-and-duplicatecorrelation" class="anchor"></a>Compare p-values from dream and duplicateCorrelation</h2>
<p>In order to understand the empircal difference between dream and duplication correlation, we can plot the <span class="math inline">\(-\log_{10}\)</span> p-values from both methods.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Compare p-values and make plot</span>
<span class="va">p1</span> <span class="op">=</span> <span class="fu"><a href="../reference/topTable.html">topTable</a></span><span class="op">(</span><span class="va">fitDupCor</span>, coef<span class="op">=</span><span class="st">"Disease1"</span>, number<span class="op">=</span><span class="cn">Inf</span>, sort.by<span class="op">=</span><span class="st">"none"</span><span class="op">)</span><span class="op">$</span><span class="va">P.Value</span>
<span class="va">p2</span> <span class="op">=</span> <span class="fu"><a href="../reference/topTable.html">topTable</a></span><span class="op">(</span><span class="va">fitmm</span>, number<span class="op">=</span><span class="cn">Inf</span>, sort.by<span class="op">=</span><span class="st">"none"</span><span class="op">)</span><span class="op">$</span><span class="va">P.Value</span>

<span class="fu"><a href="../reference/plotCompareP.html">plotCompareP</a></span><span class="op">(</span> <span class="va">p1</span>, <span class="va">p2</span>, <span class="va">vp</span><span class="op">$</span><span class="va">Individual</span>, <span class="va">dupcor</span><span class="op">$</span><span class="va">consensus</span><span class="op">)</span></code></pre></div>
<p><img src="dream_files/figure-html/define-1.png" width="700"></p>
<p>The duplicateCorrelation method estimates a single variance term genome-wide even though the donor contribution of a particular gene can vary substantially from the genome-wide trend. Using a single value genome-wide for the within-donor variance can reduce power and increase the false positive rate in a particular, reproducible way. Let <span class="math inline">\(\tau^2_g\)</span> be the value of the donor component for gene <span class="math inline">\(g\)</span> and <span class="math inline">\(\bar{\tau}^2\)</span> be the genome-wide mean. For genes where <span class="math inline">\(\tau^2_g&gt;\bar{\tau}^2\)</span>, using <span class="math inline">\(\bar{\tau}^2\)</span> under-corrects for the donor component so that it increases the false positive rate compared to using <span class="math inline">\(\tau^2_g\)</span>. Conversely, for genes where <span class="math inline">\(\tau^2_g&lt;\bar{\tau}^2\)</span>, using <span class="math inline">\(\bar{\tau}^2\)</span> over-corrects for the donor component so that it decreases power. Increasing sample size does not overcome this issue. The dream method overcomes this issue by using a <span class="math inline">\(\tau^2_g\)</span>.</p>
<p>Here, the <span class="math inline">\(-\log_{10}\)</span> p-values from both methods are plotted and colored by the donor contribution estiamted by variancePartition. The green value indicates <span class="math inline">\(\bar{\tau}^2\)</span>, while red and blue indicate higher and lower values, respectively. When only one variance component is used and the contrast matrix is simple, the effect of using dream versus duplicateCorrelation is determined by the comparison of <span class="math inline">\(\tau^2_g\)</span> to <span class="math inline">\(\bar{\tau}^2\)</span>:</p>
<p>dream can increase the <span class="math inline">\(-\log_{10}\)</span> p-value for genes with a lower donor component (i.e. <span class="math inline">\(\tau^2_g&lt;\bar{\tau}^2\)</span>) and decrease <span class="math inline">\(-\log_{10}\)</span> p-value for genes with a higher donor component (i.e. <span class="math inline">\(\tau^2_g&gt;\bar{\tau}^2\)</span>)</p>
<p>Note that using more variance components or a more complicated contrast matrix can make the relationship more complicated.</p>
</div>
</div>
<div id="parallel-processing" class="section level1">
<h1 class="hasAnchor">
<a href="#parallel-processing" class="anchor"></a>Parallel processing</h1>
<p>variancePartition functions including <code><a href="../reference/dream.html">dream()</a></code>, <code><a href="../reference/fitExtractVarPartModel.html">fitExtractVarPartModel()</a></code> and <code><a href="../reference/fitVarPartModel.html">fitVarPartModel()</a></code> can take advange of multicore machines to speed up analysis. It uses the <a href="http://bioconductor.org/packages/BiocParallel/">BiocParallel</a> package to manage the parallelization.</p>
<p>There are multiple ways to use parallel processing depending on your needs</p>
<ul>
<li>Specify parameters with the BPPARAM argument.</li>
</ul>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Request 4 cores, and enable the progress bar</span>
<span class="co"># This is the ideal for Linux, OS X and Windows</span>
<span class="va">param</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html">SnowParam</a></span><span class="op">(</span><span class="fl">4</span>, <span class="st">"SOCK"</span>, progressbar<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="va">fitmm</span> <span class="op">=</span> <span class="fu"><a href="../reference/dream.html">dream</a></span><span class="op">(</span> <span class="va">vobjDream</span>, <span class="va">form</span>, <span class="va">metadata</span>, BPPARAM<span class="op">=</span><span class="va">param</span><span class="op">)</span>

<span class="co"># Or disable parallel processing and just do serial </span>
<span class="va">param</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html">SerialParam</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">fitmm</span> <span class="op">=</span> <span class="fu"><a href="../reference/dream.html">dream</a></span><span class="op">(</span> <span class="va">vobjDream</span>, <span class="va">form</span>, <span class="va">metadata</span>, BPPARAM<span class="op">=</span><span class="va">param</span><span class="op">)</span></code></pre></div>
<ul>
<li>Set parameters globally for the entire R session</li>
</ul>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">param</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html">SnowParam</a></span><span class="op">(</span><span class="fl">4</span>, <span class="st">"SOCK"</span>, progressbar<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/register.html">register</a></span><span class="op">(</span><span class="va">param</span><span class="op">)</span>

<span class="co"># uses global parallel processing settings</span>
<span class="va">fitmm</span> <span class="op">=</span> <span class="fu"><a href="../reference/dream.html">dream</a></span><span class="op">(</span> <span class="va">vobjDream</span>, <span class="va">form</span>, <span class="va">metadata</span> <span class="op">)</span></code></pre></div>
<p>By default <code>BPPARAM</code> and the global setttings are set the results of <code><a href="https://rdrr.io/pkg/BiocParallel/man/register.html">bpparam()</a></code>. But note that using <code><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html">SnowParam()</a></code> can <em>dramatically</em> reduce the memory usage needed for parallel processing because it reduces memory redundancy between threads.</p>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h1>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R version 4.0.3 (2020-10-10)
## Platform: x86_64-apple-darwin19.6.0 (64-bit)
## Running under: macOS Catalina 10.15.7
## 
## Matrix products: default
## BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
## LAPACK: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libLAPACK.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] parallel  stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] edgeR_3.32.1             variancePartition_1.21.3 Biobase_2.50.0          
## [4] BiocGenerics_0.36.0      scales_1.1.1             BiocParallel_1.24.1     
## [7] limma_3.46.0             ggplot2_3.3.3            pander_0.6.3            
## 
## loaded via a namespace (and not attached):
##  [1] nlme_3.1-152        bitops_1.0-6        fs_1.5.0            pbkrtest_0.5.1     
##  [5] doParallel_1.0.16   progress_1.2.2      rprojroot_2.0.2     numDeriv_2016.8-1.1
##  [9] tools_4.0.3         backports_1.2.1     bslib_0.2.4         utf8_1.2.1         
## [13] R6_2.5.0            KernSmooth_2.23-18  DBI_1.1.1           colorspace_2.0-0   
## [17] withr_2.4.1         tidyselect_1.1.0    prettyunits_1.1.1   compiler_4.0.3     
## [21] textshaping_0.3.3   desc_1.3.0          labeling_0.4.2      sass_0.3.1         
## [25] caTools_1.18.1      pkgdown_1.6.1       systemfonts_1.0.1   stringr_1.4.0      
## [29] digest_0.6.27       minqa_1.2.4         rmarkdown_2.7       colorRamps_2.3     
## [33] pkgconfig_2.0.3     htmltools_0.5.1.1   lme4_1.1-26         highr_0.8          
## [37] fastmap_1.1.0       rlang_0.4.10        jquerylib_0.1.3     generics_0.1.0     
## [41] farver_2.1.0        jsonlite_1.7.2      gtools_3.8.2        dplyr_1.0.5        
## [45] magrittr_2.0.1      Matrix_1.3-2        Rcpp_1.0.6          munsell_0.5.0      
## [49] fansi_0.4.2         lifecycle_1.0.0     stringi_1.5.3       yaml_2.2.1         
## [53] MASS_7.3-53.1       gplots_3.1.1        plyr_1.8.6          grid_4.0.3         
## [57] crayon_1.4.1        lattice_0.20-41     splines_4.0.3       hms_1.0.0          
## [61] locfit_1.5-9.4      knitr_1.31          pillar_1.5.1        boot_1.3-27        
## [65] reshape2_1.4.4      codetools_0.2-18    glue_1.4.2          evaluate_0.14      
## [69] vctrs_0.3.6         nloptr_1.2.2.2      foreach_1.5.1       gtable_0.3.0       
## [73] purrr_0.3.4         tidyr_1.1.3         assertthat_0.2.1    cachem_1.0.4       
## [77] xfun_0.22           broom_0.7.5         ragg_1.1.2          tibble_3.1.0       
## [81] snow_0.4-3          lmerTest_3.1-3      iterators_1.0.13    memoise_2.0.0      
## [85] statmod_1.4.35      ellipsis_0.3.1</code></pre>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Law2014" class="csl-entry">
Law, C. W., Y. Chen, W. Shi, and G. K. Smyth. 2014. <span>“<span class="nocase">Voom: precision weights unlock linear model analysis tools for RNA-seq read counts.</span>”</span> <em>Genome Biology</em> 15 (2): R29. <a href="https://doi.org/10.1186/gb-2014-15-2-r29">https://doi.org/10.1186/gb-2014-15-2-r29</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Gabriel Hoffman.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
