<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="variancePartition">
<title>Multivariate tests â€¢ variancePartition</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Multivariate tests">
<meta property="og:description" content="variancePartition">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">variancePartition</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.28.8</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/FAQ.html">Frequently asked questions</a>
    <a class="dropdown-item" href="../articles/additional_visualization.html">Additional visualizations of variance structure</a>
    <a class="dropdown-item" href="../articles/dream.html">dream</a>
    <a class="dropdown-item" href="../articles/multivariate_tests.html">Multivariate tests</a>
    <a class="dropdown-item" href="../articles/theory_practice_random_effects.html">Theory and practice of random effects and REML</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/DiseaseNeuroGenomics/variancePartition/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Multivariate tests</h1>
            <h3 data-toc-skip class="subtitle">Combining results of
univariate tests</h3>
                        <h4 data-toc-skip class="author">Developed by <a href="http://gabrielhoffman.github.io/" class="external-link">Gabriel Hoffman</a>
</h4>
            
            <h4 data-toc-skip class="date">Run on 2023-03-08
14:18:55</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/DiseaseNeuroGenomics/variancePartition/blob/HEAD/vignettes/multivariate_tests.Rmd" class="external-link"><code>vignettes/multivariate_tests.Rmd</code></a></small>
      <div class="d-none name"><code>multivariate_tests.Rmd</code></div>
    </div>

    
    
<!---

cd /Users/gabrielhoffman/workspace/repos/variancePartition/vignettes

rmarkdown::render('multivariate_tests.Rmd')

--->
<style>
body {
text-align: justify}
</style>
<p>Results from the univariate regressions performed using can be
combined in a post-processing step to perform multivariate hypothesis
testing. In this example, we fit on transcript-level counts and then
perform multivariate hypothesis testing by combining transcripts at the
gene-level. This is done with the function.</p>
<div class="section level2">
<h2 id="import-transcript-level-counts">Import transcript-level counts<a class="anchor" aria-label="anchor" href="#import-transcript-level-counts"></a>
</h2>
<p>Read in transcript counts from the package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org" class="external-link">readr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mikelove/tximport" class="external-link">tximport</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">tximportData</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># specify directory</span></span>
<span><span class="va">path</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package<span class="op">=</span><span class="st">"tximportData"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># read sample meta-data</span></span>
<span><span class="va">samples</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">path</span>,<span class="st">"samples.txt"</span><span class="op">)</span>, header<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">samples.ext</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">path</span>,<span class="st">"samples_extended.txt"</span><span class="op">)</span>, header<span class="op">=</span><span class="cn">TRUE</span>, sep<span class="op">=</span><span class="st">"\t"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># read assignment of transcripts to genes</span></span>
<span><span class="co"># remove genes on the PAR, since these are present twice</span></span>
<span><span class="va">tx2gene</span> <span class="op">=</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"tx2gene.gencode.v27.csv"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">tx2gene</span> <span class="op">=</span> <span class="va">tx2gene</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"PAR_Y"</span>, <span class="va">tx2gene</span><span class="op">$</span><span class="va">GENEID</span>, invert<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># read transcript-level quatifictions</span></span>
<span><span class="va">files</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">path</span>, <span class="st">"salmon"</span>, <span class="va">samples</span><span class="op">$</span><span class="va">run</span>, <span class="st">"quant.sf.gz"</span><span class="op">)</span></span>
<span><span class="va">txi</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/tximport/man/tximport.html" class="external-link">tximport</a></span><span class="op">(</span><span class="va">files</span>, type <span class="op">=</span> <span class="st">"salmon"</span>, txOut<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create metadata simulating two conditions</span></span>
<span><span class="va">sampleTable</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>condition <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sampleTable</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Sample"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="standard-dream-analysis">Standard dream analysis<a class="anchor" aria-label="anchor" href="#standard-dream-analysis"></a>
</h2>
<p>Perform standard analysis at the transcript-level</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/variancePartition" class="external-link">variancePartition</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/edgeR" class="external-link">edgeR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Prepare transcript-level reads</span></span>
<span><span class="va">dge</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html" class="external-link">DGEList</a></span><span class="op">(</span><span class="va">txi</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">condition</span>, data <span class="op">=</span> <span class="va">sampleTable</span><span class="op">)</span></span>
<span><span class="va">isexpr</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html" class="external-link">filterByExpr</a></span><span class="op">(</span><span class="va">dge</span>, <span class="va">design</span><span class="op">)</span></span>
<span><span class="va">dge</span> <span class="op">=</span> <span class="va">dge</span><span class="op">[</span><span class="va">isexpr</span>,<span class="op">]</span></span>
<span><span class="va">dge</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html" class="external-link">calcNormFactors</a></span><span class="op">(</span><span class="va">dge</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Estimate precision weights</span></span>
<span><span class="va">vobj</span> <span class="op">=</span> <span class="fu"><a href="../reference/voomWithDreamWeights.html">voomWithDreamWeights</a></span><span class="op">(</span><span class="va">dge</span>, <span class="op">~</span> <span class="va">condition</span>, <span class="va">sampleTable</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Fit regression model one transcript at a time</span></span>
<span><span class="va">fit</span> <span class="op">=</span> <span class="fu"><a href="../reference/dream-method.html">dream</a></span><span class="op">(</span><span class="va">vobj</span>, <span class="op">~</span> <span class="va">condition</span>, <span class="va">sampleTable</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="multivariate-analysis">Multivariate analysis<a class="anchor" aria-label="anchor" href="#multivariate-analysis"></a>
</h2>
<p>Combine the transcript-level results at the gene-level. The mapping
between transcript and gene is stored in as a list.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Prepare transcript to gene mapping</span></span>
<span><span class="co"># keep only transcripts present in vobj</span></span>
<span><span class="co"># then convert to list with key GENEID and values TXNAMEs</span></span>
<span><span class="va">keep</span> <span class="op">=</span> <span class="va">tx2gene</span><span class="op">$</span><span class="va">TXNAME</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">vobj</span><span class="op">)</span></span>
<span><span class="va">tx2gene.lst</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/stack.html" class="external-link">unstack</a></span><span class="op">(</span><span class="va">tx2gene</span><span class="op">[</span><span class="va">keep</span>,<span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run multivariate test on entries in each feature set</span></span>
<span><span class="va">res</span> <span class="op">=</span> <span class="fu"><a href="../reference/mvTest-method.html">mvTest</a></span><span class="op">(</span><span class="va">fit</span>, <span class="va">vobj</span>, <span class="va">tx2gene.lst</span>, coef<span class="op">=</span><span class="st">"conditionB"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># truncate gene names since they have version numbers</span></span>
<span><span class="co"># ENST00000498289.5 -&gt; ENST00000498289</span></span>
<span><span class="va">res</span><span class="op">$</span><span class="va">ID.short</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"\\..+"</span>, <span class="st">""</span>, <span class="va">res</span><span class="op">$</span><span class="va">ID</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="gene-set-analysis">Gene set analysis<a class="anchor" aria-label="anchor" href="#gene-set-analysis"></a>
</h2>
<p>Perform gene set analysis using on the gene-level test
statistics.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># must have zenith &gt; v1.0.2</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://DiseaseNeuroGenomics.github.io/zenith" class="external-link">zenith</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GSEABase</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gs</span> <span class="op">=</span> <span class="fu"><a href="http://gabrielhoffman.github.io/zenith/reference/get_MSigDB.html" class="external-link">get_MSigDB</a></span><span class="op">(</span><span class="st">"C1"</span>, to<span class="op">=</span><span class="st">"ENSEMBL"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">df_gsa</span> <span class="op">=</span> <span class="fu"><a href="http://gabrielhoffman.github.io/zenith/reference/zenithPR_gsa.html" class="external-link">zenithPR_gsa</a></span><span class="op">(</span> <span class="va">res</span><span class="op">$</span><span class="va">stat</span>, <span class="va">res</span><span class="op">$</span><span class="va">ID.short</span>, <span class="va">gs</span>, inter.gene.cor<span class="op">=</span><span class="fl">.05</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">df_gsa</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                 NGenes Correlation      delta       se       p.less p.greater       PValue</span></span>
<span><span class="co">## M652_chr5q35        80        0.05 -9677.9679 1690.916 5.312232e-09 1.0000000 1.062446e-08</span></span>
<span><span class="co">## M12945_chr12q21     35        0.05 -1506.2061 1897.392 2.136533e-01 0.7863467 4.273066e-01</span></span>
<span><span class="co">## M5824_chr11p13      30        0.05  -988.0640 1952.255 3.063917e-01 0.6936083 6.127835e-01</span></span>
<span><span class="co">## M1556_chr20q11      92        0.05   775.2711 1678.107 6.779531e-01 0.3220469 6.440938e-01</span></span>
<span><span class="co">## M13160_chr16p11    106        0.05  -514.2259 1659.081 3.783030e-01 0.6216970 7.566060e-01</span></span>
<span><span class="co">## M6742_chr2q36       21        0.05  -545.5361 2108.153 3.979063e-01 0.6020937 7.958126e-01</span></span>
<span><span class="co">##                 Direction          FDR</span></span>
<span><span class="co">## M652_chr5q35         Down 2.656116e-06</span></span>
<span><span class="co">## M12945_chr12q21      Down 9.968906e-01</span></span>
<span><span class="co">## M5824_chr11p13       Down 9.968906e-01</span></span>
<span><span class="co">## M1556_chr20q11         Up 9.968906e-01</span></span>
<span><span class="co">## M13160_chr16p11      Down 9.968906e-01</span></span>
<span><span class="co">## M6742_chr2q36        Down 9.968906e-01</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<details><pre><code><span><span class="co">## R version 4.2.0 (2022-04-22)</span></span>
<span><span class="co">## Platform: x86_64-apple-darwin19.6.0 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Big Sur/Monterey 10.16</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] org.Hs.eg.db_3.15.0      msigdbr_7.5.1            GSEABase_1.58.0         </span></span>
<span><span class="co">##  [4] graph_1.74.0             annotate_1.74.0          XML_3.99-0.13           </span></span>
<span><span class="co">##  [7] AnnotationDbi_1.58.0     IRanges_2.30.1           S4Vectors_0.34.0        </span></span>
<span><span class="co">## [10] Biobase_2.58.0           BiocGenerics_0.44.0      zenith_1.0.2            </span></span>
<span><span class="co">## [13] edgeR_3.38.4             variancePartition_1.28.8 BiocParallel_1.32.0     </span></span>
<span><span class="co">## [16] limma_3.54.0             ggplot2_3.4.1            tximportData_1.24.0     </span></span>
<span><span class="co">## [19] tximport_1.24.0          readr_2.1.3             </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] backports_1.4.1             BiocFileCache_2.4.0         systemfonts_1.0.4          </span></span>
<span><span class="co">##   [4] plyr_1.8.8                  remaCor_0.0.11              splines_4.2.0              </span></span>
<span><span class="co">##   [7] GenomeInfoDb_1.32.4         digest_0.6.30               foreach_1.5.2              </span></span>
<span><span class="co">##  [10] htmltools_0.5.3             fansi_1.0.4                 magrittr_2.0.3             </span></span>
<span><span class="co">##  [13] memoise_2.0.1               doParallel_1.0.17           aod_1.3.2                  </span></span>
<span><span class="co">##  [16] tzdb_0.3.0                  Biostrings_2.64.1           matrixStats_0.63.0         </span></span>
<span><span class="co">##  [19] vroom_1.6.0                 pkgdown_2.0.6               prettyunits_1.1.1          </span></span>
<span><span class="co">##  [22] colorspace_2.1-0            rappdirs_0.3.3              blob_1.2.3                 </span></span>
<span><span class="co">##  [25] textshaping_0.3.6           rbibutils_2.2.10            xfun_0.35                  </span></span>
<span><span class="co">##  [28] dplyr_1.0.10                crayon_1.5.2                RCurl_1.98-1.9             </span></span>
<span><span class="co">##  [31] jsonlite_1.8.4              lme4_1.1-31                 iterators_1.0.14           </span></span>
<span><span class="co">##  [34] glue_1.6.2                  gtable_0.3.1                zlibbioc_1.42.0            </span></span>
<span><span class="co">##  [37] XVector_0.36.0              DelayedArray_0.22.0         EnrichmentBrowser_2.26.0   </span></span>
<span><span class="co">##  [40] Rgraphviz_2.40.0            RcppZiggurat_0.1.6          scales_1.2.1               </span></span>
<span><span class="co">##  [43] mvtnorm_1.1-3               DBI_1.1.3                   Rcpp_1.0.9                 </span></span>
<span><span class="co">##  [46] xtable_1.8-4                progress_1.2.2              bit_4.0.5                  </span></span>
<span><span class="co">##  [49] httr_1.4.4                  gplots_3.1.3                ellipsis_0.3.2             </span></span>
<span><span class="co">##  [52] pkgconfig_2.0.3             sass_0.4.4                  dbplyr_2.2.1               </span></span>
<span><span class="co">##  [55] locfit_1.5-9.6              utf8_1.2.3                  tidyselect_1.2.0           </span></span>
<span><span class="co">##  [58] rlang_1.0.6                 reshape2_1.4.4              munsell_0.5.0              </span></span>
<span><span class="co">##  [61] tools_4.2.0                 cachem_1.0.6                cli_3.6.0                  </span></span>
<span><span class="co">##  [64] generics_0.1.3              RSQLite_2.2.19              broom_1.0.1                </span></span>
<span><span class="co">##  [67] evaluate_0.18               stringr_1.5.0               fastmap_1.1.0              </span></span>
<span><span class="co">##  [70] yaml_2.3.6                  ragg_1.2.4                  babelgene_22.9             </span></span>
<span><span class="co">##  [73] RhpcBLASctl_0.21-247.1      knitr_1.41                  bit64_4.0.5                </span></span>
<span><span class="co">##  [76] fs_1.5.2                    caTools_1.18.2              purrr_0.3.5                </span></span>
<span><span class="co">##  [79] KEGGREST_1.36.3             nlme_3.1-160                KEGGgraph_1.56.0           </span></span>
<span><span class="co">##  [82] compiler_4.2.0              pbkrtest_0.5.1              filelock_1.0.2             </span></span>
<span><span class="co">##  [85] curl_4.3.3                  png_0.1-8                   clusterGeneration_1.3.7    </span></span>
<span><span class="co">##  [88] tibble_3.1.8                bslib_0.4.1                 stringi_1.7.8              </span></span>
<span><span class="co">##  [91] desc_1.4.2                  lattice_0.20-45             Matrix_1.5-3               </span></span>
<span><span class="co">##  [94] nloptr_2.0.3                vctrs_0.5.2                 pillar_1.8.1               </span></span>
<span><span class="co">##  [97] lifecycle_1.0.3             RUnit_0.4.32                Rdpack_2.4                 </span></span>
<span><span class="co">## [100] jquerylib_0.1.4             bitops_1.0-7                GenomicRanges_1.48.0       </span></span>
<span><span class="co">## [103] R6_2.5.1                    KernSmooth_2.23-20          codetools_0.2-18           </span></span>
<span><span class="co">## [106] boot_1.3-28.1               MASS_7.3-58.1               gtools_3.9.4               </span></span>
<span><span class="co">## [109] assertthat_0.2.1            SummarizedExperiment_1.26.1 rprojroot_2.0.3            </span></span>
<span><span class="co">## [112] withr_2.5.0                 GenomeInfoDbData_1.2.8      parallel_4.2.0             </span></span>
<span><span class="co">## [115] hms_1.1.2                   grid_4.2.0                  tidyr_1.2.1                </span></span>
<span><span class="co">## [118] minqa_1.2.5                 rmarkdown_2.18              Rfast_2.0.6                </span></span>
<span><span class="co">## [121] MatrixGenerics_1.8.1</span></span></code></pre>
<p>&lt;&gt;</p>
</details>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
