<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Empirical Bayes Fully Moderated t-statistics — eBayesFMT • variancePartition</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Empirical Bayes Fully Moderated t-statistics — eBayesFMT" />
<meta property="og:description" content="Empirical Bayes Fully Moderated t-statistics from linear mixed model fit with dream" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">variancePartition</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.21.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/FAQ.html">Frequently asked questions</a>
    </li>
    <li>
      <a href="../articles/additional_visualization.html">Additional visualizations of variance structure</a>
    </li>
    <li>
      <a href="../articles/dream.html">dream: Differential expression testing with linear mixed models for repeated measures</a>
    </li>
    <li>
      <a href="../articles/theory_practice_random_effects.html">Theory and practice of random effects and REML</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/GabrielHoffman/variancePartition/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Empirical Bayes Fully Moderated t-statistics</h1>
    <small class="dont-index">Source: <a href='https://github.com/GabrielHoffman/variancePartition/blob/master/R/eBayesFMT.R'><code>R/eBayesFMT.R</code></a></small>
    <div class="hidden name"><code>eBayesFMT.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Empirical Bayes Fully Moderated t-statistics from linear mixed model fit with dream</p>
    </div>

    <pre class="usage"><span class='fu'>eBayesFMT</span><span class='op'>(</span><span class='va'>fit</span>, <span class='va'>data</span>, <span class='va'>Individual</span>, method <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"VC"</span>, <span class='st'>"WS"</span><span class='op'>)</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>fit</th>
      <td><p>model fit returned by dream of class MArrayLM2</p></td>
    </tr>
    <tr>
      <th>data</th>
      <td><p>data.frame with columns corresponding to formula</p></td>
    </tr>
    <tr>
      <th>Individual</th>
      <td><p>string referring to column in data.frame indicating which individual the repeated measures come from</p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p>Use either variance components ('VC') or Welch-Satterthwaite ('WS')</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Applies empirical Bayes method of, Yu, et al. (2019) for linear mixed models.  This method applies a prior and shrinkage to 1) the residual variance and 2) the variance component estimates for 'Individual'.  'Individual' refers to the variable indicating which individual the repeated measures come from.  This method then combines posterior values from (1) and (2) to approximate the degrees of freedom for the t-statistic.</p>
<p>Yu, L., Zhang, J., Brock, G. et al. Fully moderated t-statistic in linear modeling of mixed effects for differential expression analysis. BMC Bioinformatics 20, 675 (2019). https://doi.org/10.1186/s12859-019-3248-9</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># load library</span>
<span class='co'># library(variancePartition)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://github.com/Bioconductor/BiocParallel'>BiocParallel</a></span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/pkg/BiocParallel/man/register.html'>register</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html'>SerialParam</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span>

<span class='co'># load simulated data:</span>
<span class='co'># geneExpr: matrix of gene expression values</span>
<span class='co'># info: information/metadata about each sample</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>varPartData</span><span class='op'>)</span>

<span class='va'>form</span> <span class='op'>&lt;-</span> <span class='op'>~</span> <span class='va'>Batch</span> <span class='op'>+</span> <span class='op'>(</span><span class='fl'>1</span><span class='op'>|</span><span class='va'>Individual</span><span class='op'>)</span> <span class='op'>+</span> <span class='op'>(</span><span class='fl'>1</span><span class='op'>|</span><span class='va'>Tissue</span><span class='op'>)</span> 

<span class='co'># Fit linear mixed model for each gene</span>
<span class='co'># run on just 10 genes for time</span>
<span class='va'>fit</span> <span class='op'>=</span> <span class='fu'><a href='dream.html'>dream</a></span><span class='op'>(</span> <span class='va'>geneExpr</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>10</span>,<span class='op'>]</span>, <span class='va'>form</span>, <span class='va'>info</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Dividing work into 1 chunks...</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Total:1 s</span></div><div class='input'>
<span class='co'># view top genes using standard t-statistics</span>
<span class='fu'><a href='topTable.html'>topTable</a></span><span class='op'>(</span> <span class='va'>fit</span> <span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Removing intercept from test coefficients</span></div><div class='output co'>#&gt;             Batch2       Batch3      Batch4     AveExpr          F   P.Value
#&gt; gene8   0.02567263 -0.129563777  0.67499421   0.9171386 1.96561914 0.1267816
#&gt; gene6  -0.67049689 -1.034776107 -1.04729128  -3.1554790 1.93232542 0.1319909
#&gt; gene3   0.08200976 -0.172582199 -0.49673001   0.1702122 1.44637527 0.2364513
#&gt; gene1  -0.46092425 -0.453624935 -0.05771812 -10.4664549 1.03262351 0.3833791
#&gt; gene7  -0.31606553  0.149452306 -0.03562447  -4.3799381 0.74258476 0.5301575
#&gt; gene2  -0.49420745 -0.382210977 -0.40013018  -1.1281610 0.62797976 0.5992798
#&gt; gene5   0.26795357  0.006060832  0.27368297   4.7187640 0.32958690 0.8039606
#&gt; gene9  -0.35301416 -0.096570091 -0.12646906  -2.3079042 0.27897593 0.8403964
#&gt; gene10  0.11578714 -0.098856729  0.14565869  -2.3673775 0.27305129 0.8446423
#&gt; gene4   0.11531023  0.122855760  0.13721184  -4.5359748 0.05418364 0.9832565
#&gt;        adj.P.Val      F.std
#&gt; gene8  0.6599547 1.90225556
#&gt; gene6  0.6599547 1.87125160
#&gt; gene3  0.7881710 1.41413220
#&gt; gene1  0.9384914 1.01798899
#&gt; gene7  0.9384914 0.73636776
#&gt; gene2  0.9384914 0.62417724
#&gt; gene5  0.9384914 0.32960191
#&gt; gene9  0.9384914 0.27928007
#&gt; gene10 0.9384914 0.27338242
#&gt; gene4  0.9832565 0.05449604</div><div class='input'>
<span class='co'># Compute moderated t-statistics from Yu, et al, 2019</span>
<span class='va'>fiteb</span> <span class='op'>=</span> <span class='fu'>variancePartition</span><span class='fu'>:::</span><span class='fu'>eBayesFMT</span><span class='op'>(</span><span class='va'>fit</span>, <span class='va'>info</span>, <span class='st'>'Individual'</span><span class='op'>)</span>
<span class='fu'><a href='topTable.html'>topTable</a></span><span class='op'>(</span> <span class='va'>fiteb</span> <span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Removing intercept from test coefficients</span></div><div class='output co'>#&gt; <span class='message'>topTableF is obsolete and will be removed in a future version of limma. Please considering using topTable instead.</span></div><div class='output co'>#&gt;             Batch2       Batch3      Batch4     AveExpr          F   P.Value
#&gt; gene6  -0.67049689 -1.034776107 -1.04729128  -3.1554790 0.49876352 0.6843109
#&gt; gene8   0.02567263 -0.129563777  0.67499421   0.9171386 0.36318893 0.7797634
#&gt; gene3   0.08200976 -0.172582199 -0.49673001   0.1702122 0.20303106 0.8939750
#&gt; gene7  -0.31606553  0.149452306 -0.03562447  -4.3799381 0.15424030 0.9266352
#&gt; gene1  -0.46092425 -0.453624935 -0.05771812 -10.4664549 0.14981701 0.9294858
#&gt; gene2  -0.49420745 -0.382210977 -0.40013018  -1.1281610 0.11010298 0.9539356
#&gt; gene5   0.26795357  0.006060832  0.27368297   4.7187640 0.06850536 0.9765104
#&gt; gene9  -0.35301416 -0.096570091 -0.12646906  -2.3079042 0.06256423 0.9793859
#&gt; gene10  0.11578714 -0.098856729  0.14565869  -2.3673775 0.04373560 0.9877395
#&gt; gene4   0.11531023  0.122855760  0.13721184  -4.5359748 0.01288501 0.9979822
#&gt;        adj.P.Val      F.std
#&gt; gene6  0.9979822 0.49705501
#&gt; gene8  0.9979822 0.36295415
#&gt; gene3  0.9979822 0.20357178
#&gt; gene7  0.9979822 0.15480749
#&gt; gene1  0.9979822 0.15038173
#&gt; gene2  0.9979822 0.11060913
#&gt; gene5  0.9979822 0.06887981
#&gt; gene9  0.9979822 0.06291398
#&gt; gene10 0.9979822 0.04399733
#&gt; gene4  0.9979822 0.01297045</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Gabriel Hoffman.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


