<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Empirical Bayes Fully Moderated t-statistics from linear mixed model fit with dream"><title>Empirical Bayes Fully Moderated t-statistics — eBayesFMT • variancePartition</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Empirical Bayes Fully Moderated t-statistics — eBayesFMT"><meta property="og:description" content="Empirical Bayes Fully Moderated t-statistics from linear mixed model fit with dream"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">variancePartition</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.27.5</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/FAQ.html">Frequently asked questions</a>
    <a class="dropdown-item" href="../articles/additional_visualization.html">Additional visualizations of variance structure</a>
    <a class="dropdown-item" href="../articles/dream.html">dream</a>
    <a class="dropdown-item" href="../articles/theory_practice_random_effects.html">Theory and practice of random effects and REML</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/DiseaseNeuroGenomics/variancePartition/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Empirical Bayes Fully Moderated t-statistics</h1>
      <small class="dont-index">Source: <a href="https://github.com/DiseaseNeuroGenomics/variancePartition/blob/HEAD/R/eBayesFMT.R" class="external-link"><code>R/eBayesFMT.R</code></a></small>
      <div class="d-none name"><code>eBayesFMT.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Empirical Bayes Fully Moderated t-statistics from linear mixed model fit with dream</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">eBayesFMT</span><span class="op">(</span><span class="va">fit</span>, <span class="va">data</span>, <span class="va">Individual</span>, method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"VC"</span>, <span class="st">"WS"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>fit</dt>
<dd><p>model fit returned by dream of class MArrayLM2</p></dd>


<dt>data</dt>
<dd><p>data.frame with columns corresponding to formula</p></dd>


<dt>Individual</dt>
<dd><p>string referring to column in data.frame indicating which individual the repeated measures come from</p></dd>


<dt>method</dt>
<dd><p>Use either variance components ('VC') or Welch-Satterthwaite ('WS')</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Applies empirical Bayes method of, Yu, et al. (2019) for linear mixed models.  This method applies a prior and shrinkage to 1) the residual variance and 2) the variance component estimates for 'Individual'.  'Individual' refers to the variable indicating which individual the repeated measures come from.  This method then combines posterior values from (1) and (2) to approximate the degrees of freedom for the t-statistic.</p>
<p>Yu, L., Zhang, J., Brock, G. et al. Fully moderated t-statistic in linear modeling of mixed effects for differential expression analysis. BMC Bioinformatics 20, 675 (2019). https://doi.org/10.1186/s12859-019-3248-9</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># load library</span></span></span>
<span class="r-in"><span><span class="co"># library(variancePartition)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># load simulated data:</span></span></span>
<span class="r-in"><span><span class="co"># geneExpr: matrix of gene expression values</span></span></span>
<span class="r-in"><span><span class="co"># info: information/metadata about each sample</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">varPartData</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">form</span> <span class="op">&lt;-</span> <span class="op">~</span> <span class="va">Batch</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span><span class="op">|</span><span class="va">Individual</span><span class="op">)</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span><span class="op">|</span><span class="va">Tissue</span><span class="op">)</span> </span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Fit linear mixed model for each gene</span></span></span>
<span class="r-in"><span><span class="co"># run on just 10 genes for time</span></span></span>
<span class="r-in"><span><span class="va">fit</span> <span class="op">=</span> <span class="fu"><a href="dream-method.html">dream</a></span><span class="op">(</span> <span class="va">geneExpr</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="op">]</span>, <span class="va">form</span>, <span class="va">info</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Dividing work into 1 chunks...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Total:0.6 s</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># view top genes using standard t-statistics</span></span></span>
<span class="r-in"><span><span class="fu"><a href="toptable-method.html">topTable</a></span><span class="op">(</span> <span class="va">fit</span> <span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removing intercept from test coefficients</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             Batch2       Batch3      Batch4     AveExpr          F   P.Value</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene8   0.02567263 -0.129563777  0.67499421   0.9171386 1.96561914 0.1267816</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene6  -0.67049689 -1.034776107 -1.04729128  -3.1554790 1.93232542 0.1319909</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene3   0.08200976 -0.172582199 -0.49673001   0.1702122 1.44637527 0.2364513</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene1  -0.46092425 -0.453624935 -0.05771812 -10.4664549 1.03262351 0.3833791</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene7  -0.31606553  0.149452306 -0.03562447  -4.3799381 0.74258476 0.5301575</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene2  -0.49420745 -0.382210977 -0.40013018  -1.1281610 0.62797976 0.5992798</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene5   0.26795357  0.006060832  0.27368297   4.7187640 0.32958690 0.8039606</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene9  -0.35301416 -0.096570091 -0.12646906  -2.3079042 0.27897593 0.8403964</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene10  0.11578714 -0.098856729  0.14565869  -2.3673775 0.27305129 0.8446423</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene4   0.11531023  0.122855760  0.13721184  -4.5359748 0.05418364 0.9832565</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        adj.P.Val      F.std</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene8  0.6599547 1.90225556</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene6  0.6599547 1.87125160</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene3  0.7881710 1.41413220</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene1  0.9384914 1.01798899</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene7  0.9384914 0.73636776</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene2  0.9384914 0.62417724</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene5  0.9384914 0.32960191</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene9  0.9384914 0.27928007</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene10 0.9384914 0.27338242</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene4  0.9832565 0.05449604</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Compute moderated t-statistics from Yu, et al, 2019</span></span></span>
<span class="r-in"><span><span class="va">fiteb</span> <span class="op">=</span> <span class="fu">variancePartition</span><span class="fu">:::</span><span class="fu">eBayesFMT</span><span class="op">(</span><span class="va">fit</span>, <span class="va">info</span>, <span class="st">'Individual'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="toptable-method.html">topTable</a></span><span class="op">(</span> <span class="va">fiteb</span> <span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removing intercept from test coefficients</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> topTableF is obsolete and will be removed in a future version of limma. Please considering using topTable instead.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             Batch2       Batch3      Batch4     AveExpr          F   P.Value</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene6  -0.67049689 -1.034776107 -1.04729128  -3.1554790 0.49876352 0.6843109</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene8   0.02567263 -0.129563777  0.67499421   0.9171386 0.36318893 0.7797634</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene3   0.08200976 -0.172582199 -0.49673001   0.1702122 0.20303106 0.8939750</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene7  -0.31606553  0.149452306 -0.03562447  -4.3799381 0.15424030 0.9266352</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene1  -0.46092425 -0.453624935 -0.05771812 -10.4664549 0.14981701 0.9294858</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene2  -0.49420745 -0.382210977 -0.40013018  -1.1281610 0.11010298 0.9539356</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene5   0.26795357  0.006060832  0.27368297   4.7187640 0.06850536 0.9765104</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene9  -0.35301416 -0.096570091 -0.12646906  -2.3079042 0.06256423 0.9793859</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene10  0.11578714 -0.098856729  0.14565869  -2.3673775 0.04373560 0.9877395</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene4   0.11531023  0.122855760  0.13721184  -4.5359748 0.01288501 0.9979822</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        adj.P.Val      F.std</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene6  0.9979822 0.49705501</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene8  0.9979822 0.36295415</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene3  0.9979822 0.20357178</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene7  0.9979822 0.15480749</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene1  0.9979822 0.15038173</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene2  0.9979822 0.11060913</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene5  0.9979822 0.06887981</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene9  0.9979822 0.06291398</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene10 0.9979822 0.04399733</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> gene4  0.9979822 0.01297045</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by <a href="http://gabrielhoffman.github.io" class="external-link">Gabriel Hoffman</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

