<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Fit linear (mixed) model — fitVarPartModel • variancePartition</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Fit linear (mixed) model — fitVarPartModel" />
<meta property="og:description" content="Fit linear (mixed) model to estimate contribution of multiple sources of variation while simultaneously correcting for all other variables." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">variancePartition</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.21.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/FAQ.html">Frequently asked questions</a>
    </li>
    <li>
      <a href="../articles/additional_visualization.html">Additional visualizations of variance structure</a>
    </li>
    <li>
      <a href="../articles/dream.html">dream: Differential expression testing with linear mixed models for repeated measures</a>
    </li>
    <li>
      <a href="../articles/theory_practice_random_effects.html">Theory and practice of random effects and REML</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/GabrielHoffman/variancePartition/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Fit linear (mixed) model</h1>
    <small class="dont-index">Source: <a href='https://github.com/GabrielHoffman/variancePartition/blob/master/R/fitModels.R'><code>R/fitModels.R</code></a></small>
    <div class="hidden name"><code>fitVarPartModel-method.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Fit linear (mixed) model to estimate contribution of multiple sources of variation while simultaneously correcting for all other variables.</p>
    </div>

    <pre class="usage"><span class='fu'>fitVarPartModel</span><span class='op'>(</span>
  <span class='va'>exprObj</span>,
  <span class='va'>formula</span>,
  <span class='va'>data</span>,
  REML <span class='op'>=</span> <span class='cn'>FALSE</span>,
  useWeights <span class='op'>=</span> <span class='cn'>TRUE</span>,
  weightsMatrix <span class='op'>=</span> <span class='cn'>NULL</span>,
  showWarnings <span class='op'>=</span> <span class='cn'>TRUE</span>,
  fxn <span class='op'>=</span> <span class='va'>identity</span>,
  control <span class='op'>=</span> <span class='fu'>lme4</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/lme4/man/lmerControl.html'>lmerControl</a></span><span class='op'>(</span>calc.derivs <span class='op'>=</span> <span class='cn'>FALSE</span>, check.rankX <span class='op'>=</span> <span class='st'>"stop.deficient"</span><span class='op'>)</span>,
  quiet <span class='op'>=</span> <span class='cn'>FALSE</span>,
  BPPARAM <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/BiocParallel/man/register.html'>bpparam</a></span><span class='op'>(</span><span class='op'>)</span>,
  <span class='va'>...</span>
<span class='op'>)</span>

<span class='co'># S4 method for matrix</span>
<span class='fu'>fitVarPartModel</span><span class='op'>(</span>
  <span class='va'>exprObj</span>,
  <span class='va'>formula</span>,
  <span class='va'>data</span>,
  REML <span class='op'>=</span> <span class='cn'>FALSE</span>,
  useWeights <span class='op'>=</span> <span class='cn'>TRUE</span>,
  weightsMatrix <span class='op'>=</span> <span class='cn'>NULL</span>,
  showWarnings <span class='op'>=</span> <span class='cn'>TRUE</span>,
  fxn <span class='op'>=</span> <span class='va'>identity</span>,
  control <span class='op'>=</span> <span class='fu'>lme4</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/lme4/man/lmerControl.html'>lmerControl</a></span><span class='op'>(</span>calc.derivs <span class='op'>=</span> <span class='cn'>FALSE</span>, check.rankX <span class='op'>=</span> <span class='st'>"stop.deficient"</span><span class='op'>)</span>,
  quiet <span class='op'>=</span> <span class='cn'>FALSE</span>,
  BPPARAM <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/BiocParallel/man/register.html'>bpparam</a></span><span class='op'>(</span><span class='op'>)</span>,
  <span class='va'>...</span>
<span class='op'>)</span>

<span class='co'># S4 method for data.frame</span>
<span class='fu'>fitVarPartModel</span><span class='op'>(</span>
  <span class='va'>exprObj</span>,
  <span class='va'>formula</span>,
  <span class='va'>data</span>,
  REML <span class='op'>=</span> <span class='cn'>FALSE</span>,
  useWeights <span class='op'>=</span> <span class='cn'>TRUE</span>,
  weightsMatrix <span class='op'>=</span> <span class='cn'>NULL</span>,
  showWarnings <span class='op'>=</span> <span class='cn'>TRUE</span>,
  fxn <span class='op'>=</span> <span class='va'>identity</span>,
  control <span class='op'>=</span> <span class='fu'>lme4</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/lme4/man/lmerControl.html'>lmerControl</a></span><span class='op'>(</span>calc.derivs <span class='op'>=</span> <span class='cn'>FALSE</span>, check.rankX <span class='op'>=</span> <span class='st'>"stop.deficient"</span><span class='op'>)</span>,
  quiet <span class='op'>=</span> <span class='cn'>FALSE</span>,
  BPPARAM <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/BiocParallel/man/register.html'>bpparam</a></span><span class='op'>(</span><span class='op'>)</span>,
  <span class='va'>...</span>
<span class='op'>)</span>

<span class='co'># S4 method for EList</span>
<span class='fu'>fitVarPartModel</span><span class='op'>(</span>
  <span class='va'>exprObj</span>,
  <span class='va'>formula</span>,
  <span class='va'>data</span>,
  REML <span class='op'>=</span> <span class='cn'>FALSE</span>,
  useWeights <span class='op'>=</span> <span class='cn'>TRUE</span>,
  weightsMatrix <span class='op'>=</span> <span class='cn'>NULL</span>,
  showWarnings <span class='op'>=</span> <span class='cn'>TRUE</span>,
  fxn <span class='op'>=</span> <span class='va'>identity</span>,
  control <span class='op'>=</span> <span class='fu'>lme4</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/lme4/man/lmerControl.html'>lmerControl</a></span><span class='op'>(</span>calc.derivs <span class='op'>=</span> <span class='cn'>FALSE</span>, check.rankX <span class='op'>=</span> <span class='st'>"stop.deficient"</span><span class='op'>)</span>,
  quiet <span class='op'>=</span> <span class='cn'>FALSE</span>,
  BPPARAM <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/BiocParallel/man/register.html'>bpparam</a></span><span class='op'>(</span><span class='op'>)</span>,
  <span class='va'>...</span>
<span class='op'>)</span>

<span class='co'># S4 method for ExpressionSet</span>
<span class='fu'>fitVarPartModel</span><span class='op'>(</span>
  <span class='va'>exprObj</span>,
  <span class='va'>formula</span>,
  <span class='va'>data</span>,
  REML <span class='op'>=</span> <span class='cn'>FALSE</span>,
  useWeights <span class='op'>=</span> <span class='cn'>TRUE</span>,
  weightsMatrix <span class='op'>=</span> <span class='cn'>NULL</span>,
  showWarnings <span class='op'>=</span> <span class='cn'>TRUE</span>,
  fxn <span class='op'>=</span> <span class='va'>identity</span>,
  control <span class='op'>=</span> <span class='fu'>lme4</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/lme4/man/lmerControl.html'>lmerControl</a></span><span class='op'>(</span>calc.derivs <span class='op'>=</span> <span class='cn'>FALSE</span>, check.rankX <span class='op'>=</span> <span class='st'>"stop.deficient"</span><span class='op'>)</span>,
  quiet <span class='op'>=</span> <span class='cn'>FALSE</span>,
  BPPARAM <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/BiocParallel/man/register.html'>bpparam</a></span><span class='op'>(</span><span class='op'>)</span>,
  <span class='va'>...</span>
<span class='op'>)</span>

<span class='co'># S4 method for sparseMatrix</span>
<span class='fu'>fitVarPartModel</span><span class='op'>(</span>
  <span class='va'>exprObj</span>,
  <span class='va'>formula</span>,
  <span class='va'>data</span>,
  REML <span class='op'>=</span> <span class='cn'>FALSE</span>,
  useWeights <span class='op'>=</span> <span class='cn'>TRUE</span>,
  weightsMatrix <span class='op'>=</span> <span class='cn'>NULL</span>,
  showWarnings <span class='op'>=</span> <span class='cn'>TRUE</span>,
  fxn <span class='op'>=</span> <span class='va'>identity</span>,
  control <span class='op'>=</span> <span class='fu'>lme4</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/lme4/man/lmerControl.html'>lmerControl</a></span><span class='op'>(</span>calc.derivs <span class='op'>=</span> <span class='cn'>FALSE</span>, check.rankX <span class='op'>=</span> <span class='st'>"stop.deficient"</span><span class='op'>)</span>,
  quiet <span class='op'>=</span> <span class='cn'>FALSE</span>,
  BPPARAM <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/BiocParallel/man/register.html'>bpparam</a></span><span class='op'>(</span><span class='op'>)</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>exprObj</th>
      <td><p>matrix of expression data (g genes x n samples), or <code>ExpressionSet</code>, or <code>EList</code> returned by <code>voom()</code> from the <code>limma</code> package</p></td>
    </tr>
    <tr>
      <th>formula</th>
      <td><p>specifies variables for the linear (mixed) model.  Must only specify covariates, since the rows of exprObj are automatically used a a response. e.g.: <code>~ a + b + (1|c)</code></p></td>
    </tr>
    <tr>
      <th>data</th>
      <td><p><code>data.frame</code> with columns corresponding to formula</p></td>
    </tr>
    <tr>
      <th>REML</th>
      <td><p>use restricted maximum likelihood to fit linear mixed model. default is FALSE.  See Details.</p></td>
    </tr>
    <tr>
      <th>useWeights</th>
      <td><p>if TRUE, analysis uses heteroskedastic error estimates from voom().  Value is ignored unless exprObj is an <code>EList()</code> from <code>voom()</code> or <code>weightsMatrix</code> is specified</p></td>
    </tr>
    <tr>
      <th>weightsMatrix</th>
      <td><p>matrix the same dimension as exprObj with observation-level weights from <code>voom()</code>.  Used only if useWeights is TRUE</p></td>
    </tr>
    <tr>
      <th>showWarnings</th>
      <td><p>show warnings about model fit (default TRUE)</p></td>
    </tr>
    <tr>
      <th>fxn</th>
      <td><p>apply function to model fit for each gene.  Defaults to identify function so it returns the model fit itself</p></td>
    </tr>
    <tr>
      <th>control</th>
      <td><p>control settings for <code><a href='https://rdrr.io/pkg/lme4/man/lmer.html'>lmer()</a></code></p></td>
    </tr>
    <tr>
      <th>quiet</th>
      <td><p>suppress message, default FALSE</p></td>
    </tr>
    <tr>
      <th>BPPARAM</th>
      <td><p>parameters for parallel evaluation</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments for <code><a href='https://rdrr.io/pkg/lme4/man/lmer.html'>lmer()</a></code> or <code><a href='https://rdrr.io/r/stats/lm.html'>lm()</a></code></p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code><a href='https://rdrr.io/r/base/list.html'>list()</a></code> of where each entry is a model fit produced by <code><a href='https://rdrr.io/pkg/lme4/man/lmer.html'>lmer()</a></code> or <code><a href='https://rdrr.io/r/stats/lm.html'>lm()</a></code></p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>A linear (mixed) model is fit for each gene in exprObj, using formula to specify variables in the regression.  If categorical variables are modeled as random effects (as is recommended), then a linear mixed model us used.  For example if formula is <code>~ a + b + (1|c)</code>, then the model is</p>
<p><code>fit &lt;- lmer( exprObj[j,] ~ a + b + (1|c), data=data)</code></p>
<p>If there are no random effects, so formula is <code>~ a + b + c</code>, a 'standard' linear model is used:</p>
<p><code>fit &lt;- lm( exprObj[j,] ~ a + b + c, data=data)</code></p>
<p>In both cases, <code>useWeights=TRUE</code> causes <code>weightsMatrix[j,]</code> to be included as weights in the regression model.</p>
<p>Note: Fitting the model for 20,000 genes can be computationally intensive.  To accelerate computation, models can be fit in parallel using <code>BiocParallel</code> to run in parallel.  Parallel processing must be enabled before calling this function.  See below.</p>
<p>The regression model is fit for each gene separately. Samples with missing values in either gene expression or metadata are omitted by the underlying call to lm/lmer.</p>
<p>Since this function returns a list of each model fit, using this function is slower and uses more memory than <code><a href='fitExtractVarPartModel.html'>fitExtractVarPartModel()</a></code>.</p>
<p><code>REML=FALSE</code> uses maximum likelihood to estimate variance fractions.  This approach produced unbiased estimates, while <code>REML=TRUE</code> can show substantial bias.  See Vignette "3) Theory and practice of random effects and REML"</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='co'># load library</span>
<span class='co'># library(variancePartition)</span>

<span class='co'># Intialize parallel backend with 4 cores</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://github.com/Bioconductor/BiocParallel'>BiocParallel</a></span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/pkg/BiocParallel/man/register.html'>register</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html'>SnowParam</a></span><span class='op'>(</span><span class='fl'>4</span><span class='op'>)</span><span class='op'>)</span>

<span class='co'># load simulated data:</span>
<span class='co'># geneExpr: matrix of gene expression values</span>
<span class='co'># info: information/metadata about each sample</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>varPartData</span><span class='op'>)</span>

<span class='co'># Specify variables to consider</span>
<span class='co'># Age is continuous so we model it as a fixed effect</span>
<span class='co'># Individual and Tissue are both categorical, so we model them as random effects</span>
<span class='va'>form</span> <span class='op'>&lt;-</span> <span class='op'>~</span> <span class='va'>Age</span> <span class='op'>+</span> <span class='op'>(</span><span class='fl'>1</span><span class='op'>|</span><span class='va'>Individual</span><span class='op'>)</span> <span class='op'>+</span> <span class='op'>(</span><span class='fl'>1</span><span class='op'>|</span><span class='va'>Tissue</span><span class='op'>)</span> 

<span class='co'># Step 1: fit linear mixed model on gene expression</span>
<span class='co'># If categorical variables are specified, a linear mixed model is used</span>
<span class='co'># If all variables are modeled as continuous, a linear model is used</span>
<span class='co'># each entry in results is a regression model fit on a single gene</span>
<span class='co'># Step 2: extract variance fractions from each model fit</span>
<span class='co'># for each gene, returns fraction of variation attributable to each variable </span>
<span class='co'># Interpretation: the variance explained by each variable</span>
<span class='co'># after correction for all other variables</span>
<span class='va'>varPart</span> <span class='op'>&lt;-</span> <span class='fu'><a href='fitExtractVarPartModel.html'>fitExtractVarPartModel</a></span><span class='op'>(</span> <span class='va'>geneExpr</span>, <span class='va'>form</span>, <span class='va'>info</span> <span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Dividing work into 10 chunks...</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Total:14 s</span></div><div class='input'> 
<span class='co'># violin plot of contribution of each variable to total variance</span>
<span class='co'># also sort columns</span>
<span class='fu'><a href='plotVarPart.html'>plotVarPart</a></span><span class='op'>(</span> <span class='fu'><a href='sortCols.html'>sortCols</a></span><span class='op'>(</span> <span class='va'>varPart</span> <span class='op'>)</span> <span class='op'>)</span>
</div><div class='img'><img src='fitVarPartModel-method-1.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'># Advanced: </span>
<span class='co'># Fit model and extract variance in two separate steps</span>
<span class='co'># Step 1: fit model for each gene, store model fit for each gene in a list</span>
<span class='va'>results</span> <span class='op'>&lt;-</span> <span class='fu'>fitVarPartModel</span><span class='op'>(</span> <span class='va'>geneExpr</span>, <span class='va'>form</span>, <span class='va'>info</span> <span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Memory usage to store result: &gt;8.9 Mb</span></div><div class='output co'>#&gt; <span class='message'>Dividing work into 10 chunks...</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Total:19 s</span></div><div class='input'>
<span class='co'># Step 2: extract variance fractions</span>
<span class='va'>varPart</span> <span class='op'>&lt;-</span> <span class='fu'><a href='extractVarPart.html'>extractVarPart</a></span><span class='op'>(</span> <span class='va'>results</span> <span class='op'>)</span>

<span class='co'># Note: fitVarPartModel also accepts ExpressionSet</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>sample.ExpressionSet</span>, package<span class='op'>=</span><span class='st'>"Biobase"</span><span class='op'>)</span>

<span class='co'># ExpressionSet example</span>
<span class='va'>form</span> <span class='op'>&lt;-</span> <span class='op'>~</span> <span class='op'>(</span><span class='fl'>1</span><span class='op'>|</span><span class='va'>sex</span><span class='op'>)</span> <span class='op'>+</span> <span class='op'>(</span><span class='fl'>1</span><span class='op'>|</span><span class='va'>type</span><span class='op'>)</span> <span class='op'>+</span> <span class='va'>score</span>
<span class='va'>info2</span> <span class='op'>&lt;-</span> <span class='fu'>Biobase</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/Biobase/man/phenoData.html'>pData</a></span><span class='op'>(</span><span class='va'>sample.ExpressionSet</span><span class='op'>)</span>
<span class='va'>results2</span> <span class='op'>&lt;-</span> <span class='fu'>fitVarPartModel</span><span class='op'>(</span> <span class='va'>sample.ExpressionSet</span>, <span class='va'>form</span>, <span class='va'>info2</span> <span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Memory usage to store result: &gt;15.2 Mb</span></div><div class='output co'>#&gt; <span class='message'>Dividing work into 25 chunks...</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Total:131 s</span></div><div class='input'>

</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Gabriel Hoffman.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


