<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Construct Matrix of Custom Contrasts — makeContrastsDream • variancePartition</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Construct Matrix of Custom Contrasts — makeContrastsDream" />
<meta property="og:description" content="Construct the contrast matrix corresponding to specified contrasts of a set of parameters." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">variancePartition</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.21.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/FAQ.html">Frequently asked questions</a>
    </li>
    <li>
      <a href="../articles/additional_visualization.html">Additional visualizations of variance structure</a>
    </li>
    <li>
      <a href="../articles/dream.html">dream: Differential expression testing with linear mixed models for repeated measures</a>
    </li>
    <li>
      <a href="../articles/theory_practice_random_effects.html">Theory and practice of random effects and REML</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/GabrielHoffman/variancePartition/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Construct Matrix of Custom Contrasts</h1>
    <small class="dont-index">Source: <a href='https://github.com/GabrielHoffman/variancePartition/blob/master/R/contrasts.R'><code>R/contrasts.R</code></a></small>
    <div class="hidden name"><code>makeContrastsDream.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Construct the contrast matrix corresponding to specified contrasts of a set of parameters.</p>
    </div>

    <pre class="usage"><span class='fu'>makeContrastsDream</span><span class='op'>(</span><span class='va'>formula</span>, <span class='va'>data</span>, <span class='va'>...</span>, contrasts <span class='op'>=</span> <span class='cn'>NULL</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>formula</th>
      <td><p>specifies variables for the linear (mixed) model.  Must only specify covariates, since the rows of exprObj are automatically used a a response. e.g.: <code>~ a + b + (1|c)</code>  Formulas with only fixed effects also work</p></td>
    </tr>
    <tr>
      <th>data</th>
      <td><p>data.frame with columns corresponding to formula</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>expressions, or character strings which can be parsed to expressions, specifying contrasts</p></td>
    </tr>
    <tr>
      <th>contrasts</th>
      <td><p>character vector specifying contrasts</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>matrix of linear contrasts between regression coefficients</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function expresses contrasts between a set of parameters as a numeric matrix. The parameters are usually the coefficients from a linear (mixed) model fit, so the matrix specifies which comparisons between the coefficients are to be extracted from the fit. The output from this function is usually used as input to <code><a href='dream.html'>dream()</a></code>.</p>
<p>This function is inspired by <code><a href='https://rdrr.io/pkg/limma/man/makeContrasts.html'>limma::makeContrasts()</a></code> but is designed to be compatible with linear mixed models for <code><a href='dream.html'>dream()</a></code></p>
<p>Names in ... and contrasts will be used as column names in the returned value.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># load library</span>
<span class='co'># library(variancePartition)</span>

<span class='co'># Intialize parallel backend with 4 cores</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://github.com/Bioconductor/BiocParallel'>BiocParallel</a></span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/pkg/BiocParallel/man/register.html'>register</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html'>SnowParam</a></span><span class='op'>(</span><span class='fl'>4</span><span class='op'>)</span><span class='op'>)</span>

<span class='co'># load simulated data:</span>
<span class='co'># geneExpr: matrix of gene expression values</span>
<span class='co'># info: information/metadata about each sample</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>varPartData</span><span class='op'>)</span>

<span class='va'>form</span> <span class='op'>&lt;-</span> <span class='op'>~</span> <span class='fl'>0</span> <span class='op'>+</span> <span class='va'>Batch</span> <span class='op'>+</span> <span class='op'>(</span><span class='fl'>1</span><span class='op'>|</span><span class='va'>Individual</span><span class='op'>)</span> <span class='op'>+</span> <span class='op'>(</span><span class='fl'>1</span><span class='op'>|</span><span class='va'>Tissue</span><span class='op'>)</span> 

<span class='co'># Define contrasts</span>
<span class='va'>L</span> <span class='op'>=</span> <span class='fu'>makeContrastsDream</span><span class='op'>(</span> <span class='va'>form</span>, <span class='va'>info</span>, contrasts <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span>Batch1_vs_2 <span class='op'>=</span> <span class='st'>"Batch1 - Batch2"</span>, Batch3_vs_4 <span class='op'>=</span> <span class='st'>"Batch3 - Batch4"</span>, Batch1_vs_34 <span class='op'>=</span> <span class='st'>"Batch1 - (Batch3 + Batch4)/2"</span><span class='op'>)</span><span class='op'>)</span>

<span class='co'># show contrasts matrix</span>
<span class='va'>L</span>
</div><div class='output co'>#&gt;         Contrasts
#&gt; Levels   Batch1_vs_2 Batch3_vs_4 Batch1_vs_34
#&gt;   Batch1           1           0          1.0
#&gt;   Batch2          -1           0          0.0
#&gt;   Batch3           0           1         -0.5
#&gt;   Batch4           0          -1         -0.5</div><div class='input'>
<span class='co'># Plot to visualize contrasts matrix</span>
<span class='fu'><a href='plotContrasts.html'>plotContrasts</a></span><span class='op'>(</span><span class='va'>L</span><span class='op'>)</span>
</div><div class='img'><img src='makeContrastsDream-1.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'># Fit linear mixed model for each gene</span>
<span class='co'># run on just 10 genes for time</span>
<span class='va'>fit</span> <span class='op'>=</span> <span class='fu'><a href='dream.html'>dream</a></span><span class='op'>(</span> <span class='va'>geneExpr</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>10</span>,<span class='op'>]</span>, <span class='va'>form</span>, <span class='va'>info</span>, L<span class='op'>=</span><span class='va'>L</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Dividing work into 1 chunks...</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Total:5 s</span></div><div class='input'>
<span class='co'># examine contrasts after fitting</span>
<span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/stats/coef.html'>coef</a></span><span class='op'>(</span><span class='va'>fit</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='output co'>#&gt;       Batch1_vs_2 Batch3_vs_4 Batch1_vs_34      Batch1      Batch2      Batch3
#&gt; gene1  0.46092425 -0.39590682    0.2556715 -10.1696108 -10.6305350 -10.6232357
#&gt; gene2  0.49420745  0.01791921    0.3911706  -0.7408132  -1.2350206  -1.1230241
#&gt; gene3 -0.08200976  0.32414781    0.3346561   0.3469564   0.4289662   0.1743742
#&gt; gene4 -0.11531023 -0.01435608   -0.1300338  -4.6207874  -4.5054771  -4.4979316
#&gt; gene5 -0.26795357 -0.26762214   -0.1398719   4.6243196   4.8922731   4.6303804
#&gt; gene6  0.67049689  0.01251517    1.0410337  -2.4959563  -3.1664532  -3.5307325
#&gt;            Batch4
#&gt; gene1 -10.2273289
#&gt; gene2  -1.1409433
#&gt; gene3  -0.1497736
#&gt; gene4  -4.4835755
#&gt; gene5   4.8980025
#&gt; gene6  -3.5432476</div><div class='input'>
<span class='co'># show results from first contrast</span>
<span class='fu'><a href='topTable.html'>topTable</a></span><span class='op'>(</span><span class='va'>fit</span>, coef<span class='op'>=</span><span class='st'>"Batch1_vs_2"</span><span class='op'>)</span>
</div><div class='output co'>#&gt;              logFC     AveExpr           t   P.Value adj.P.Val       z.std
#&gt; gene6   0.67049689  -3.1554790  1.52366339 0.1319712 0.6497599  1.50637414
#&gt; gene2   0.49420745  -1.1281610  1.36219015 0.1774115 0.6497599  1.34876797
#&gt; gene1   0.46092425 -10.4664549  1.30835356 0.1949280 0.6497599  1.29613790
#&gt; gene7   0.31606553  -4.3799381  0.84073462 0.4032967 0.8431315  0.83574794
#&gt; gene9   0.35301610  -2.3079042  0.77799477 0.4391161 0.8431315  0.77368663
#&gt; gene5  -0.26795357   4.7187640 -0.66862559 0.5058789 0.8431315 -0.66526828
#&gt; gene10 -0.11578714  -2.3673775 -0.35846423 0.7210510 0.8940896 -0.35705451
#&gt; gene4  -0.11531023  -4.5359748 -0.34244225 0.7330176 0.8940896 -0.34111426
#&gt; gene3  -0.08200976   0.1702122 -0.24821747 0.8046806 0.8940896 -0.24729409
#&gt; gene8  -0.02567263   0.9171386 -0.06725961 0.9465626 0.9465626 -0.06702401</div><div class='input'>
<span class='co'># show results from second contrast</span>
<span class='fu'><a href='topTable.html'>topTable</a></span><span class='op'>(</span><span class='va'>fit</span>, coef<span class='op'>=</span><span class='st'>"Batch3_vs_4"</span><span class='op'>)</span>
</div><div class='output co'>#&gt;              logFC     AveExpr           t    P.Value adj.P.Val       z.std
#&gt; gene8  -0.80455799   0.9171386 -2.16904635 0.03340121 0.3340121 -2.12722716
#&gt; gene1  -0.39590682 -10.4664549 -1.16045670 0.24967624 0.9767482 -1.15113612
#&gt; gene3   0.32414781   0.1702122  1.01160906 0.31512596 0.9767482  1.00452431
#&gt; gene10 -0.24451542  -2.3673775 -0.77856133 0.43880768 0.9767482 -0.77420820
#&gt; gene5  -0.26762214   4.7187640 -0.68664313 0.49452493 0.9767482 -0.68312963
#&gt; gene7   0.18507678  -4.3799381  0.50649399 0.61406680 0.9767482  0.50427691
#&gt; gene9   0.02990086  -2.3079042  0.06771983 0.94619681 0.9767482  0.06748348
#&gt; gene2   0.01791921  -1.1281610  0.05082868 0.95960356 0.9767482  0.05065108
#&gt; gene4  -0.01435608  -4.5359748 -0.04385867 0.96513877 0.9767482 -0.04370599
#&gt; gene6   0.01251517  -3.1554790  0.02924752 0.97674821 0.9767482  0.02914593</div><div class='input'>
<span class='co'># show results from third contrast</span>
<span class='fu'><a href='topTable.html'>topTable</a></span><span class='op'>(</span><span class='va'>fit</span>, coef<span class='op'>=</span><span class='st'>"Batch1_vs_34"</span><span class='op'>)</span>
</div><div class='output co'>#&gt;              logFC     AveExpr           t    P.Value adj.P.Val       z.std
#&gt; gene6   1.04103369  -3.1554790  2.37923408 0.01998146 0.1998146  2.32669579
#&gt; gene2   0.39117058  -1.1281610  1.08409908 0.28193902 0.9421343  1.07597371
#&gt; gene3   0.33465610   0.1702122  1.01819011 0.31201734 0.9421343  1.01099809
#&gt; gene1   0.25567153 -10.4664549  0.72973043 0.46791597 0.9421343  0.72587408
#&gt; gene8  -0.27271522   0.9171386 -0.71831187 0.47489082 0.9421343 -0.71454406
#&gt; gene4  -0.13003380  -4.5359748 -0.38839218 0.69886587 0.9421343 -0.38685188
#&gt; gene5  -0.13987190   4.7187640 -0.35098159 0.72662357 0.9421343 -0.34962044
#&gt; gene9   0.11152288  -2.3079042  0.24726631 0.80539881 0.9421343  0.24636615
#&gt; gene7  -0.05691392  -4.3799381 -0.15221534 0.87944287 0.9421343 -0.15167551
#&gt; gene10 -0.02340098  -2.3673775 -0.07284137 0.94213432 0.9421343 -0.07258756</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Gabriel Hoffman.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


