<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Transform RNA-Seq Data Ready for Linear Mixed Modelling with dream() — voomWithDreamWeights • variancePartition</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Transform RNA-Seq Data Ready for Linear Mixed Modelling with dream() — voomWithDreamWeights" />
<meta property="og:description" content="Transform count data to log2-counts per million (logCPM), estimate the mean-variance relationship and use this to compute appropriate observation-level weights. The data are then ready for linear mixed modelling with dream().   This method is the same as limma::voom(), except that it allows random effects in the formula" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">variancePartition</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.25.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/FAQ.html">Frequently asked questions</a>
    </li>
    <li>
      <a href="../articles/additional_visualization.html">Additional visualizations of variance structure</a>
    </li>
    <li>
      <a href="../articles/dream.html">dream: Differential expression testing with linear mixed models for repeated measures</a>
    </li>
    <li>
      <a href="../articles/theory_practice_random_effects.html">Theory and practice of random effects and REML</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/GabrielHoffman/variancePartition/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Transform RNA-Seq Data Ready for Linear Mixed Modelling with <code>dream()</code></h1>
    <small class="dont-index">Source: <a href='https://github.com/GabrielHoffman/variancePartition/blob/master/R/voomWithDreamWeights.R'><code>R/voomWithDreamWeights.R</code></a></small>
    <div class="hidden name"><code>voomWithDreamWeights.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Transform count data to log2-counts per million (logCPM), estimate the mean-variance relationship and use this to compute appropriate observation-level weights. The data are then ready for linear mixed modelling with <code><a href='dream.html'>dream()</a></code>.   This method is the same as limma::voom(), except that it allows random effects in the formula</p>
    </div>

    <pre class="usage"><span class='fu'>voomWithDreamWeights</span><span class='op'>(</span>
  <span class='va'>counts</span>,
  <span class='va'>formula</span>,
  <span class='va'>data</span>,
  lib.size <span class='op'>=</span> <span class='cn'>NULL</span>,
  normalize.method <span class='op'>=</span> <span class='st'>"none"</span>,
  span <span class='op'>=</span> <span class='fl'>0.5</span>,
  weights <span class='op'>=</span> <span class='cn'>NULL</span>,
  plot <span class='op'>=</span> <span class='cn'>FALSE</span>,
  save.plot <span class='op'>=</span> <span class='cn'>FALSE</span>,
  quiet <span class='op'>=</span> <span class='cn'>FALSE</span>,
  BPPARAM <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/BiocParallel/man/register.html'>bpparam</a></span><span class='op'>(</span><span class='op'>)</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>counts</th>
      <td><p>a numeric <code>matrix</code> containing raw counts, or an <code>ExpressionSet</code> containing raw counts, or a <code>DGEList</code> object. Counts must be non-negative and NAs are not permitted.</p></td>
    </tr>
    <tr>
      <th>formula</th>
      <td><p>specifies variables for the linear (mixed) model.  Must only specify covariates, since the rows of exprObj are automatically used a a response. e.g.: <code>~ a + b + (1|c)</code>  Formulas with only fixed effects also work, and <code>lmFit()</code> followed by contrasts.fit() are run.</p></td>
    </tr>
    <tr>
      <th>data</th>
      <td><p><code>data.frame</code> with columns corresponding to formula</p></td>
    </tr>
    <tr>
      <th>lib.size</th>
      <td><p>numeric vector containing total library sizes for each sample.  Defaults to the normalized (effective) library sizes in <code>counts</code> if <code>counts</code> is a <code>DGEList</code> or to the columnwise count totals if <code>counts</code> is a matrix.</p></td>
    </tr>
    <tr>
      <th>normalize.method</th>
      <td><p>the microarray-style normalization method to be applied to the logCPM values (if any).  Choices are as for the <code>method</code> argument of <code>normalizeBetweenArrays</code> when the data is single-channel.  Any normalization factors found in <code>counts</code> will still be used even if <code>normalize.method="none"</code>.</p></td>
    </tr>
    <tr>
      <th>span</th>
      <td><p>width of the lowess smoothing window as a proportion.</p></td>
    </tr>
    <tr>
      <th>weights</th>
      <td><p>Can be a numeric matrix of individual weights of same dimensions as the <code>counts</code>, or a numeric vector of sample weights with length equal to <code><a href='https://rdrr.io/r/base/nrow.html'>ncol(counts)</a></code></p></td>
    </tr>
    <tr>
      <th>plot</th>
      <td><p>logical, should a plot of the mean-variance trend be displayed?</p></td>
    </tr>
    <tr>
      <th>save.plot</th>
      <td><p>logical, should the coordinates and line of the plot be saved in the output?</p></td>
    </tr>
    <tr>
      <th>quiet</th>
      <td><p>suppress message, default FALSE</p></td>
    </tr>
    <tr>
      <th>BPPARAM</th>
      <td><p>parameters for parallel evaluation</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>other arguments are passed to <code>lmer</code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>An <code>EList</code> object just like the result of <code><a href='https://rdrr.io/pkg/limma/man/voom.html'>limma::voom()</a></code></p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Adapted from <code>vomm()</code> in <code>limma</code> v3.40.2</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>limma::voom()</p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># library(variancePartition)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='http://bioinf.wehi.edu.au/edgeR'>edgeR</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://github.com/Bioconductor/BiocParallel'>BiocParallel</a></span><span class='op'>)</span>

<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>varPartDEdata</span><span class='op'>)</span>

<span class='co'># normalize RNA-seq counts</span>
<span class='va'>dge</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/edgeR/man/DGEList.html'>DGEList</a></span><span class='op'>(</span>counts <span class='op'>=</span> <span class='va'>countMatrix</span><span class='op'>)</span>
<span class='va'>dge</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/pkg/edgeR/man/calcNormFactors.html'>calcNormFactors</a></span><span class='op'>(</span><span class='va'>dge</span><span class='op'>)</span>

<span class='co'># specify formula with random effect for Individual</span>
<span class='va'>form</span> <span class='op'>&lt;-</span> <span class='op'>~</span> <span class='va'>Disease</span> <span class='op'>+</span> <span class='op'>(</span><span class='fl'>1</span><span class='op'>|</span><span class='va'>Individual</span><span class='op'>)</span> 

<span class='co'># compute observation weights</span>
<span class='va'>vobj</span> <span class='op'>=</span> <span class='fu'>voomWithDreamWeights</span><span class='op'>(</span> <span class='va'>dge</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>20</span>,<span class='op'>]</span>, <span class='va'>form</span>, <span class='va'>metadata</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Memory usage to store result: &gt;49.7 Kb</span></div><div class='output co'>#&gt; <span class='message'>Dividing work into 1 chunks...</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Total:4 s</span></div><div class='input'>
<span class='co'># fit dream model </span>
<span class='va'>res</span> <span class='op'>=</span> <span class='fu'><a href='dream.html'>dream</a></span><span class='op'>(</span> <span class='va'>vobj</span>, <span class='va'>form</span>, <span class='va'>metadata</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Dividing work into 1 chunks...</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Total:4 s</span></div><div class='input'>
<span class='co'># extract results</span>
<span class='fu'><a href='topTable.html'>topTable</a></span><span class='op'>(</span><span class='va'>res</span>, coef<span class='op'>=</span><span class='st'>"Disease1"</span><span class='op'>)</span>
</div><div class='output co'>#&gt;                                         logFC  AveExpr        t      P.Value
#&gt; ENST00000456159.1 gene=MET          1.0182945 2.458926 6.707820 9.637382e-07
#&gt; ENST00000418210.2 gene=TMEM64       1.0375652 4.715367 8.047821 4.039375e-06
#&gt; ENST00000555834.1 gene=RPS6KL1      0.9355651 5.272063 6.029943 4.544895e-06
#&gt; ENST00000248564.5 gene=GNG11        1.0543333 4.511181 5.184523 3.373837e-05
#&gt; ENST00000570099.1 gene=YPEL3        0.9654350 2.063331 7.009797 3.876342e-05
#&gt; ENST00000317802.7 gene=TSPYL6       1.0257641 4.321189 4.662013 9.772471e-04
#&gt; ENST00000482796.1 gene=RP11-101E3.5 0.9257951 1.533654 3.753019 1.099478e-03
#&gt; ENST00000267622.4 gene=TRIP11       0.6226163 5.942593 3.672330 1.336570e-03
#&gt; ENST00000337859.6 gene=ZC3H15       0.8379527 4.169523 3.552296 1.785163e-03
#&gt; ENST00000598661.1 gene=AL136376.1   0.8931155 1.231360 4.229887 1.792281e-03
#&gt;                                        adj.P.Val    z.std
#&gt; ENST00000456159.1 gene=MET          1.927476e-05 4.898902
#&gt; ENST00000418210.2 gene=TMEM64       3.029930e-05 4.609346
#&gt; ENST00000555834.1 gene=RPS6KL1      3.029930e-05 4.584769
#&gt; ENST00000248564.5 gene=GNG11        1.550537e-04 4.146644
#&gt; ENST00000570099.1 gene=YPEL3        1.550537e-04 4.114729
#&gt; ENST00000317802.7 gene=TSPYL6       3.141365e-03 3.296997
#&gt; ENST00000482796.1 gene=RP11-101E3.5 3.141365e-03 3.263751
#&gt; ENST00000267622.4 gene=TRIP11       3.341425e-03 3.208010
#&gt; ENST00000337859.6 gene=ZC3H15       3.385222e-03 3.123826
#&gt; ENST00000598661.1 gene=AL136376.1   3.385222e-03 3.122654</div><div class='input'>

</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Gabriel Hoffman.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


